// @tevm/cli/dist/commands/create.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box } from 'ink';
import { useEffect } from 'react';
import { Creating } from '../components/Creating.js';
import { FancyCreateTitle } from '../components/FancyCreateTitle.js';
import { InteractivePrompt } from '../components/InteractivePrompt.js';
import { useStore } from '../state/Store.js';
// Add command description for help output
export const description = 'Create a new Ethereum account or smart contract';
export default function Create({ options, args: [defaultName] }) {
    // Initialize store with default values
    useEffect(() => {
        useStore.setState({
            name: defaultName,
            currentStep: 0,
            path: '.',
            nameInput: '',
            framework: options.template,
            useCase: 'ui',
            packageManager: 'npm',
            noGit: false,
            noInstall: false,
            currentPage: options.skipPrompts ? 'creating' : 'interactive',
            walletConnectProjectId: '',
        });
    }, [defaultName, options.skipPrompts, options.template]);
    const store = useStore();
    const pages = {
        interactive: _jsx(InteractivePrompt, { defaultName: defaultName, store: store }),
        creating: _jsx(Creating, { store: store }),
        complete: _jsx(Box, { children: "Complete!" }),
    };
    return (_jsxs(Box, { display: "flex", flexDirection: "column", children: [_jsx(FancyCreateTitle, { loading: store.currentPage === 'creating' }, store.currentPage), pages[store.currentPage]] }));
}
