// @tevm/cli/dist/commands/getAccount.js
import { jsx as _jsx } from "react/jsx-runtime";
import { option } from 'pastel';
import { z } from 'zod';
import CliAction from '../components/CliAction.js';
import { envVar, useAction } from '../hooks/useAction.js';
// Add command description for help output
export const description = 'Get account information including balance and nonce';
// Options definitions and descriptions
const optionDescriptions = {
    address: 'Ethereum address of the account to get (env: TEVM_ADDRESS)',
    rpc: 'RPC endpoint (env: TEVM_RPC)',
    returnStorage: 'Return account storage (can be expensive) (env: TEVM_RETURN_STORAGE)',
    blockTag: 'Block tag to fetch account from (latest, earliest, pending, etc.) (env: TEVM_BLOCK_TAG)',
};
// Empty args tuple
export const args = z.tuple([]);
export const options = z.object({
    // Required address
    address: z.string().describe(option({
        description: optionDescriptions.address,
    })),
    // Interactive mode flag (run directly without interactive editor)
    run: z
        .boolean()
        .default(false)
        .describe(option({
        description: 'Run directly without interactive parameter editing (env: TEVM_RUN)',
        alias: 'r',
    })),
    // Transport options
    rpc: z
        .string()
        .default(envVar('rpc') || 'http://localhost:8545')
        .describe(option({
        description: optionDescriptions.rpc,
        defaultValueDescription: 'http://localhost:8545',
    })),
    // Account retrieval options
    returnStorage: z
        .boolean()
        .default(false)
        .describe(option({
        description: optionDescriptions.returnStorage,
    })),
    blockTag: z
        .string()
        .optional()
        .describe(option({
        description: optionDescriptions.blockTag,
    })),
    // Output formatting
    formatJson: z
        .boolean()
        .default(envVar('format_json') !== 'false')
        .describe(option({
        description: 'Format output as JSON (env: TEVM_FORMAT_JSON)',
        defaultValueDescription: 'true',
    })),
});
// Default values for all parameters
const defaultValues = {
    rpc: 'http://localhost:8545',
    returnStorage: false,
    blockTag: 'latest',
};
export default function GetAccount({ options }) {
    // Use the action hook with inlined createParams and executeAction
    const actionResult = useAction({
        actionName: 'get-account',
        options,
        defaultValues,
        optionDescriptions,
        // Inlined createParams function
        createParams: (enhancedOptions) => {
            const targetAddress = enhancedOptions['address'];
            if (!targetAddress) {
                throw new Error('Address is required');
            }
            return {
                address: targetAddress,
                returnStorage: enhancedOptions['returnStorage'] ?? false,
                blockTag: enhancedOptions['blockTag'] ?? undefined,
            };
        },
        // Inlined executeAction function
        executeAction: async (client, params) => {
            return await client.tevmGetAccount(params);
        },
    });
    // Render the action UI
    return (_jsx(CliAction, { ...actionResult, targetName: `account ${actionResult.options['address']}`, successMessage: "Account retrieved successfully!" }));
}
