// @tevm/cli/dist/commands/getChainId.js
import { jsx as _jsx } from "react/jsx-runtime";
import { option } from 'pastel';
import { z } from 'zod';
import CliAction from '../components/CliAction.js';
import { useAction } from '../hooks/useAction.js';
// Add command description for help output
export const description = 'Get the chain ID of the connected blockchain network';
// Options definitions and descriptions
const optionDescriptions = {
    rpc: 'RPC endpoint (env: TEVM_RPC)',
};
// Empty args tuple
export const args = z.tuple([]);
export const options = z.object({
    // Interactive mode flag
    run: z
        .boolean()
        .default(false)
        .describe(option({
        description: 'Run directly without interactive parameter editing (env: TEVM_RUN)',
        alias: 'r',
    })),
    // Transport options
    rpc: z
        .string()
        .optional()
        .describe(option({
        description: optionDescriptions.rpc,
        defaultValueDescription: 'http://localhost:8545',
    })),
    // Output formatting
    formatJson: z
        .boolean()
        .optional()
        .describe(option({
        description: 'Format output as JSON (env: TEVM_FORMAT_JSON)',
        defaultValueDescription: 'true',
    })),
});
// COMPREHENSIVE DEFAULTS
const defaultValues = {
    rpc: 'http://localhost:8545',
};
export default function GetChainId({ options }) {
    // Use the action hook
    const actionResult = useAction({
        actionName: 'getChainId',
        options,
        defaultValues,
        optionDescriptions,
        // Create params - getChainId takes no parameters
        createParams: (_enhancedOptions) => {
            return {};
        },
        // Execute the action
        executeAction: async (client, _params) => {
            return await client.getChainId();
        },
    });
    // If editor is active, render nothing
    if (actionResult.editorActive) {
        return null;
    }
    return _jsx(CliAction, { ...actionResult, targetName: "chain", successMessage: "Chain ID retrieved successfully!" });
}
