// @tevm/decorators/dist/index.js
import { getStorageAtHandler, getCodeHandler, getBalanceHandler, gasPriceHandler, chainIdHandler, ethCallHandler, blockNumberHandler, requestProcedure, requestBulkProcedure, loadStateHandler, dumpStateHandler, contractHandler, callHandler, setAccountHandler, getAccountHandler, mineHandler, deployHandler, dealHandler } from '@tevm/actions';
import { withRetry } from 'viem';

// src/actions/ethActions.js
var ethActions = () => (client) => {
  const wrappedEth = (() => {
    if (!("eth" in client)) {
      return {};
    }
    if (typeof client.eth !== "object") {
      throw new Error(
        "Cannot extend eth with ethActions decorator. detected a client.eth property that is not an object"
      );
    }
    return client.eth ?? {};
  })();
  return {
    eth: {
      ...wrappedEth,
      blockNumber: blockNumberHandler(client),
      call: ethCallHandler(client),
      chainId: chainIdHandler(client),
      gasPrice: gasPriceHandler(client),
      getBalance: getBalanceHandler(client),
      getCode: getCodeHandler(client),
      getStorageAt: getStorageAtHandler(client)
    }
  };
};
var getAccountAction = () => (client) => {
  return {
    getAccount: getAccountHandler(client)
  };
};
var setAccountAction = () => (client) => {
  return {
    setAccount: setAccountHandler(client)
  };
};
var callAction = () => (client) => {
  return {
    call: callHandler(client)
  };
};
var contractAction = () => (client) => {
  return {
    contract: contractHandler(client)
  };
};
var dumpStateAction = () => (client) => {
  return {
    dumpState: dumpStateHandler(client)
  };
};
var loadStateAction = () => (client) => {
  return {
    loadState: loadStateHandler(client)
  };
};
var mineAction = () => (client) => {
  return {
    mine: mineHandler(client)
  };
};
var deployAction = () => (client) => {
  return {
    deploy: deployHandler(client)
  };
};
var dealAction = () => (client) => {
  return {
    deal: dealHandler(client)
  };
};
var tevmActions = () => (client) => {
  return client.extend(loadStateAction()).extend(dumpStateAction()).extend(contractAction()).extend(callAction()).extend(setAccountAction()).extend(getAccountAction()).extend(mineAction()).extend(deployAction()).extend(dealAction());
};
var requestEip1193 = () => (client) => {
  return {
    request: async (args, options) => {
      return withRetry(async () => {
        const result = await requestProcedure(client)(
          /** @type any*/
          {
            jsonrpc: "2.0",
            id: 1,
            method: args.method,
            ...args.params ? { params: args.params } : {}
          }
        );
        if (result.error) {
          throw result.error;
        }
        return (
          /** @type {any}*/
          result.result
        );
      }, options);
    }
  };
};
var tevmSend = () => (client) => {
  return {
    send: requestProcedure(client),
    sendBulk: requestBulkProcedure(client)
  };
};

export { ethActions, requestEip1193, tevmActions, tevmSend };
