// @tevm/decorators/dist/index.cjs
'use strict';

var actions = require('@tevm/actions');
var viem = require('viem');

// src/actions/ethActions.js
var ethActions = () => (client) => {
  const wrappedEth = (() => {
    if (!("eth" in client)) {
      return {};
    }
    if (typeof client.eth !== "object") {
      throw new Error(
        "Cannot extend eth with ethActions decorator. detected a client.eth property that is not an object"
      );
    }
    return client.eth ?? {};
  })();
  return {
    eth: {
      ...wrappedEth,
      blockNumber: actions.blockNumberHandler(client),
      call: actions.ethCallHandler(client),
      chainId: actions.chainIdHandler(client),
      gasPrice: actions.gasPriceHandler(client),
      getBalance: actions.getBalanceHandler(client),
      getCode: actions.getCodeHandler(client),
      getStorageAt: actions.getStorageAtHandler(client)
    }
  };
};
var getAccountAction = () => (client) => {
  return {
    getAccount: actions.getAccountHandler(client)
  };
};
var setAccountAction = () => (client) => {
  return {
    setAccount: actions.setAccountHandler(client)
  };
};
var callAction = () => (client) => {
  return {
    call: actions.callHandler(client)
  };
};
var contractAction = () => (client) => {
  return {
    contract: actions.contractHandler(client)
  };
};
var dumpStateAction = () => (client) => {
  return {
    dumpState: actions.dumpStateHandler(client)
  };
};
var loadStateAction = () => (client) => {
  return {
    loadState: actions.loadStateHandler(client)
  };
};
var mineAction = () => (client) => {
  return {
    mine: actions.mineHandler(client)
  };
};
var deployAction = () => (client) => {
  return {
    deploy: actions.deployHandler(client)
  };
};
var dealAction = () => (client) => {
  return {
    deal: actions.dealHandler(client)
  };
};
var tevmActions = () => (client) => {
  return client.extend(loadStateAction()).extend(dumpStateAction()).extend(contractAction()).extend(callAction()).extend(setAccountAction()).extend(getAccountAction()).extend(mineAction()).extend(deployAction()).extend(dealAction());
};
var requestEip1193 = () => (client) => {
  return {
    request: async (args, options) => {
      return viem.withRetry(async () => {
        const result = await actions.requestProcedure(client)(
          /** @type any*/
          {
            jsonrpc: "2.0",
            id: 1,
            method: args.method,
            ...args.params ? { params: args.params } : {}
          }
        );
        if (result.error) {
          throw result.error;
        }
        return (
          /** @type {any}*/
          result.result
        );
      }, options);
    }
  };
};
var tevmSend = () => (client) => {
  return {
    send: actions.requestProcedure(client),
    sendBulk: actions.requestBulkProcedure(client)
  };
};

exports.ethActions = ethActions;
exports.requestEip1193 = requestEip1193;
exports.tevmActions = tevmActions;
exports.tevmSend = tevmSend;
