// @tevm/cli/dist/components/InteractivePrompt.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, useInput } from 'ink';
import { MultipleChoiceStep } from './MultipleChoiceStep.js';
import { TextInputStep } from './TextInputStep.js';
export const InteractivePrompt = ({ defaultName, store }) => {
    useInput((_, { leftArrow }) => {
        if (leftArrow) {
            store.goToPreviousStep({});
        }
    });
    const steps = [];
    steps.push(_jsx(TextInputStep, { name: "Name", isActive: store.currentStep === steps.length, color: "#A4DDED", step: {
            type: 'input',
            prompt: 'What is the name of your project?',
            stateKey: 'name',
        }, value: store.nameInput, placeholder: defaultName, onChange: (value) => {
            store.setInput({ input: 'nameInput', value });
        }, onSubmit: (value) => {
            store.selectAndContinue({ name: 'name', value: value || defaultName });
        } }, "name"));
    steps.push(_jsx(MultipleChoiceStep, { name: "Template", isActive: store.currentStep === steps.length, hide: store.currentStep < steps.length, color: "#B19CD9", multipleChoice: {
            type: 'multiple-choice',
            prompt: 'Pick a template',
            stateKey: 'framework',
            choices: {
                hardhat: { value: 'hardhat', label: 'Hardhat - Popular Ethereum development environment' },
                foundry: { value: 'foundry', label: 'Foundry - Fast, portable and modular toolkit' },
            },
        }, selectedChoice: store.framework, onSelect: (value) => {
            store.selectAndContinue({
                name: 'framework',
                value,
                nextPage: true,
            });
        } }, "template"));
    return _jsxs(Box, { flexDirection: "column", children: [...steps] });
};
