// @tevm/common/dist/index.js
import { createCustomCommon, Mainnet } from '@ethereumjs/common';
export { ConsensusAlgorithm, ConsensusType } from '@ethereumjs/common';
import { InvalidParamsError } from '@tevm/errors';
import { createLogger } from '@tevm/logger';
import { keccak256 } from '@tevm/utils';
import { defineChain, bytesToHex } from 'viem';
import { abstractTestnet as abstractTestnet$1, acala as acala$1, ancient8 as ancient8$1, ancient8Sepolia as ancient8Sepolia$1, anvil as anvil$1, apexTestnet as apexTestnet$1, arbitrum as arbitrum$1, arbitrumGoerli as arbitrumGoerli$1, arbitrumNova as arbitrumNova$1, arbitrumSepolia as arbitrumSepolia$1, areonNetwork as areonNetwork$1, areonNetworkTestnet as areonNetworkTestnet$1, artelaTestnet as artelaTestnet$1, assetChainTestnet as assetChainTestnet$1, astar as astar$1, astarZkEVM as astarZkEVM$1, astarZkyoto as astarZkyoto$1, atletaOlympia as atletaOlympia$1, aurora as aurora$1, auroraTestnet as auroraTestnet$1, auroria as auroria$1, avalanche as avalanche$1, avalancheFuji as avalancheFuji$1, b3 as b3$1, b3Sepolia as b3Sepolia$1, bahamut as bahamut$1, base as base$1, baseGoerli as baseGoerli$1, baseSepolia as baseSepolia$1, beam as beam$1, beamTestnet as beamTestnet$1, bearNetworkChainMainnet as bearNetworkChainMainnet$1, bearNetworkChainTestnet as bearNetworkChainTestnet$1, berachainTestnet as berachainTestnet$1, berachainTestnetbArtio as berachainTestnetbArtio$1, bevmMainnet as bevmMainnet$1, bitkub as bitkub$1, bitkubTestnet as bitkubTestnet$1, bitTorrent as bitTorrent$1, bitTorrentTestnet as bitTorrentTestnet$1, blast as blast$1, blastSepolia as blastSepolia$1, bob as bob$1, boba as boba$1, bobaSepolia as bobaSepolia$1, bobSepolia as bobSepolia$1, botanixTestnet as botanixTestnet$1, bronos as bronos$1, bronosTestnet as bronosTestnet$1, bsc as bsc$1, bscGreenfield as bscGreenfield$1, bscTestnet as bscTestnet$1, btr as btr$1, btrTestnet as btrTestnet$1, bxn as bxn$1, bxnTestnet as bxnTestnet$1, canto as canto$1, celo as celo$1, celoAlfajores as celoAlfajores$1, chiliz as chiliz$1, chips as chips$1, classic as classic$1, confluxESpace as confluxESpace$1, confluxESpaceTestnet as confluxESpaceTestnet$1, coreDao as coreDao$1, crab as crab$1, cronos as cronos$1, cronosTestnet as cronosTestnet$1, cronoszkEVM as cronoszkEVM$1, cronoszkEVMTestnet as cronoszkEVMTestnet$1, crossbell as crossbell$1, curtis as curtis$1, cyber as cyber$1, cyberTestnet as cyberTestnet$1, darwinia as darwinia$1, dchain as dchain$1, dchainTestnet as dchainTestnet$1, defichainEvm as defichainEvm$1, defichainEvmTestnet as defichainEvmTestnet$1, degen as degen$1, dfk as dfk$1, dodochainTestnet as dodochainTestnet$1, dogechain as dogechain$1, dreyerxMainnet as dreyerxMainnet$1, dreyerxTestnet as dreyerxTestnet$1, edgeless as edgeless$1, edgelessTestnet as edgelessTestnet$1, edgeware as edgeware$1, edgewareTestnet as edgewareTestnet$1, ekta as ekta$1, ektaTestnet as ektaTestnet$1, eon as eon$1, eos as eos$1, eosTestnet as eosTestnet$1, etherlink as etherlink$1, etherlinkTestnet as etherlinkTestnet$1, evmos as evmos$1, evmosTestnet as evmosTestnet$1, fantom as fantom$1, fantomSonicTestnet as fantomSonicTestnet$1, fantomTestnet as fantomTestnet$1, fibo as fibo$1, filecoin as filecoin$1, filecoinCalibration as filecoinCalibration$1, filecoinHyperspace as filecoinHyperspace$1, flare as flare$1, flareTestnet as flareTestnet$1, flowMainnet as flowMainnet$1, flowPreviewnet as flowPreviewnet$1, flowTestnet as flowTestnet$1, fluence as fluence$1, fluenceStage as fluenceStage$1, fluenceTestnet as fluenceTestnet$1, forma as forma$1, foundry as foundry$1, fraxtal as fraxtal$1, fraxtalTestnet as fraxtalTestnet$1, funkiMainnet as funkiMainnet$1, funkiSepolia as funkiSepolia$1, fuse as fuse$1, fuseSparknet as fuseSparknet$1, gnosis as gnosis$1, gnosisChiado as gnosisChiado$1, gobi as gobi$1, goerli as goerli$1, gravity as gravity$1, ham as ham$1, haqqMainnet as haqqMainnet$1, haqqTestedge2 as haqqTestedge2$1, hardhat as hardhat$1, harmonyOne as harmonyOne$1, hashkeyTestnet as hashkeyTestnet$1, hedera as hedera$1, hederaPreviewnet as hederaPreviewnet$1, hederaTestnet as hederaTestnet$1, holesky as holesky$1, immutableZkEvm as immutableZkEvm$1, immutableZkEvmTestnet as immutableZkEvmTestnet$1, inEVM as inEVM$1, iota as iota$1, iotaTestnet as iotaTestnet$1, iotex as iotex$1, iotexTestnet as iotexTestnet$1, jbc as jbc$1, jbcTestnet as jbcTestnet$1, kaia as kaia$1, kairos as kairos$1, kakarotSepolia as kakarotSepolia$1, karura as karura$1, kava as kava$1, kavaTestnet as kavaTestnet$1, kcc as kcc$1, klaytn as klaytn$1, klaytnBaobab as klaytnBaobab$1, koi as koi$1, kroma as kroma$1, kromaSepolia as kromaSepolia$1, l3x as l3x$1, l3xTestnet as l3xTestnet$1, lightlinkPegasus as lightlinkPegasus$1, lightlinkPhoenix as lightlinkPhoenix$1, linea as linea$1, lineaGoerli as lineaGoerli$1, lineaSepolia as lineaSepolia$1, lineaTestnet as lineaTestnet$1, lisk as lisk$1, liskSepolia as liskSepolia$1, localhost as localhost$1, lukso as lukso$1, luksoTestnet as luksoTestnet$1, lycan as lycan$1, lyra as lyra$1, mainnet as mainnet$1, mandala as mandala$1, manta as manta$1, mantaSepoliaTestnet as mantaSepoliaTestnet$1, mantaTestnet as mantaTestnet$1, mantle as mantle$1, mantleSepoliaTestnet as mantleSepoliaTestnet$1, mantleTestnet as mantleTestnet$1, merlin as merlin$1, metachain as metachain$1, metachainIstanbul as metachainIstanbul$1, metalL2 as metalL2$1, meter as meter$1, meterTestnet as meterTestnet$1, metis as metis$1, metisGoerli as metisGoerli$1, mev as mev$1, mevTestnet as mevTestnet$1, mintSepoliaTestnet as mintSepoliaTestnet$1, mode as mode$1, modeTestnet as modeTestnet$1, moonbaseAlpha as moonbaseAlpha$1, moonbeam as moonbeam$1, moonbeamDev as moonbeamDev$1, moonriver as moonriver$1, morphHolesky as morphHolesky$1, morphSepolia as morphSepolia$1, nautilus as nautilus$1, neonDevnet as neonDevnet$1, neonMainnet as neonMainnet$1, nexi as nexi$1, nexilix as nexilix$1, oasisTestnet as oasisTestnet$1, oasys as oasys$1, okc as okc$1, oortMainnetDev as oortMainnetDev$1, opBNB as opBNB$1, opBNBTestnet as opBNBTestnet$1, optimism as optimism$1, optimismGoerli as optimismGoerli$1, optimismSepolia as optimismSepolia$1, otimDevnet as otimDevnet$1, palm as palm$1, palmTestnet as palmTestnet$1, pgn as pgn$1, pgnTestnet as pgnTestnet$1, phoenix as phoenix$1, playfiAlbireo as playfiAlbireo$1, plinga as plinga$1, plumeTestnet as plumeTestnet$1, polygon as polygon$1, polygonAmoy as polygonAmoy$1, polygonMumbai as polygonMumbai$1, polygonZkEvm as polygonZkEvm$1, polygonZkEvmCardona as polygonZkEvmCardona$1, polygonZkEvmTestnet as polygonZkEvmTestnet$1, pulsechain as pulsechain$1, pulsechainV4 as pulsechainV4$1, qMainnet as qMainnet$1, qTestnet as qTestnet$1, real as real$1, redbellyTestnet as redbellyTestnet$1, redstone as redstone$1, reyaNetwork as reyaNetwork$1, rollux as rollux$1, rolluxTestnet as rolluxTestnet$1, ronin as ronin$1, root as root$1, rootPorcini as rootPorcini$1, rootstock as rootstock$1, rootstockTestnet as rootstockTestnet$1, rss3 as rss3$1, rss3Sepolia as rss3Sepolia$1, saigon as saigon$1, sapphire as sapphire$1, sapphireTestnet as sapphireTestnet$1, satoshiVM as satoshiVM$1, satoshiVMTestnet as satoshiVMTestnet$1, scroll as scroll$1, scrollSepolia as scrollSepolia$1, sei as sei$1, seiDevnet as seiDevnet$1, seiTestnet as seiTestnet$1, sepolia as sepolia$1, shapeSepolia as shapeSepolia$1, shardeumSphinx as shardeumSphinx$1, shibarium as shibarium$1, shibariumTestnet as shibariumTestnet$1, shimmer as shimmer$1, shimmerTestnet as shimmerTestnet$1, skaleBlockBrawlers as skaleBlockBrawlers$1, skaleCalypso as skaleCalypso$1, skaleCalypsoTestnet as skaleCalypsoTestnet$1, skaleCryptoBlades as skaleCryptoBlades$1, skaleCryptoColosseum as skaleCryptoColosseum$1, skaleEuropa as skaleEuropa$1, skaleEuropaTestnet as skaleEuropaTestnet$1, skaleExorde as skaleExorde$1, skaleHumanProtocol as skaleHumanProtocol$1, skaleNebula as skaleNebula$1, skaleNebulaTestnet as skaleNebulaTestnet$1, skaleRazor as skaleRazor$1, skaleTitan as skaleTitan$1, skaleTitanTestnet as skaleTitanTestnet$1, sketchpad as sketchpad$1, soneiumMinato as soneiumMinato$1, songbird as songbird$1, songbirdTestnet as songbirdTestnet$1, sophonTestnet as sophonTestnet$1, spicy as spicy$1, storyTestnet as storyTestnet$1, stratis as stratis$1, syscoin as syscoin$1, syscoinTestnet as syscoinTestnet$1, taiko as taiko$1, taikoHekla as taikoHekla$1, taikoJolnir as taikoJolnir$1, taikoKatla as taikoKatla$1, taikoTestnetSepolia as taikoTestnetSepolia$1, taraxa as taraxa$1, taraxaTestnet as taraxaTestnet$1, telcoinTestnet as telcoinTestnet$1, telos as telos$1, telosTestnet as telosTestnet$1, tenet as tenet$1, thaiChain as thaiChain$1, thunderTestnet as thunderTestnet$1, tron as tron$1, unreal as unreal$1, vechain as vechain$1, wanchain as wanchain$1, wanchainTestnet as wanchainTestnet$1, wemix as wemix$1, wemixTestnet as wemixTestnet$1, x1Testnet as x1Testnet$1, xai as xai$1, xaiTestnet as xaiTestnet$1, xdc as xdc$1, xdcTestnet as xdcTestnet$1, xLayer as xLayer$1, xLayerTestnet as xLayerTestnet$1, xrSepolia as xrSepolia$1, yooldoVerse as yooldoVerse$1, yooldoVerseTestnet as yooldoVerseTestnet$1, zetachain as zetachain$1, zetachainAthensTestnet as zetachainAthensTestnet$1, zhejiang as zhejiang$1, zilliqa as zilliqa$1, zilliqaTestnet as zilliqaTestnet$1, zircuitTestnet as zircuitTestnet$1, zkFair as zkFair$1, zkFairTestnet as zkFairTestnet$1, zkLinkNova as zkLinkNova$1, zkLinkNovaSepoliaTestnet as zkLinkNovaSepoliaTestnet$1, zksync as zksync$1, zksyncInMemoryNode as zksyncInMemoryNode$1, zksyncLocalNode as zksyncLocalNode$1, zksyncSepoliaTestnet as zksyncSepoliaTestnet$1, zora as zora$1, zoraSepolia as zoraSepolia$1, zoraTestnet as zoraTestnet$1 } from 'viem/chains';

// src/index.ts
var createMockKzg = () => {
  const mockHash = bytesToHex(keccak256("0x69", "bytes"));
  return {
    loadTrustedSetup: () => 69,
    verifyKzgProof: () => true,
    freeTrustedSetup: () => {
    },
    verifyBlobKzgProof: () => true,
    blobToKzgCommitment: () => mockHash,
    computeBlobKzgProof: () => mockHash,
    verifyBlobKzgProofBatch: () => true,
    // New methods required by the updated KZG interface
    computeBlobProof: () => mockHash,
    verifyProof: () => true,
    verifyBlobProofBatch: () => true
  };
};

// src/createCommon.js
var createCommon = ({
  customCrypto = {},
  loggingLevel = "warn",
  hardfork = "prague",
  eips = [],
  ...chain
}) => {
  try {
    const logger = createLogger({ level: loggingLevel, name: "@tevm/common" });
    const eipsArray = Array.isArray(eips) ? eips : [];
    const finalCustomCrypto = customCrypto && Object.keys(customCrypto).length > 0 ? { kzg: createMockKzg(), ...customCrypto } : { kzg: createMockKzg() };
    const ethjsCommon = createCustomCommon(
      {
        chainId: chain.id,
        name: chain.name || "TevmCustom"
      },
      Mainnet,
      {
        hardfork,
        eips: [...eipsArray, 1559, 4895, 4844, 4788, 2935],
        customCrypto: finalCustomCrypto,
        params: {
          1559: {
            elasticityMultiplier: 2,
            baseFeeMaxChangeDenominator: 8,
            initialBaseFee: 1e9
          },
          4844: {
            targetBlobGasPerBlock: 393216,
            blobGasPerBlob: 131072,
            minBlobGasPrice: 1,
            blobGasPriceUpdateFraction: 3338477
          },
          4788: {
            historicalRootsLength: 8191
          },
          // VM params accessed via param('vm', ...)
          vm: {
            historicalRootsLength: 8191,
            historyStorageAddress: "0x0aae40965e6800cd9b1f4b05ff21581047e3f91e",
            historyServeWindow: 8192
          },
          gasConfig: {
            maxRefundQuotient: 5
          }
        }
      }
    );
    if (ethjsCommon.isActivatedEIP(6800)) {
      logger.warn("verkle state is currently not supported in tevm");
    }
    logger.debug(ethjsCommon.eips(), "Created common with eips enabled");
    return {
      ...chain,
      ethjsCommon,
      copy: () => {
        const ethjsCommonCopy = ethjsCommon.copy();
        const newCommon = createCommon({ loggingLevel, hardfork, eips, ...chain });
        newCommon.ethjsCommon = ethjsCommonCopy;
        return newCommon;
      }
    };
  } catch (e) {
    const err = (
      /** @type {Error} */
      e
    );
    throw new InvalidParamsError(err.message, { cause: err });
  }
};
var abstractTestnet = createCommon({
  ...abstractTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var acala = createCommon({
  ...acala$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ancient8 = createCommon({
  ...ancient8$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ancient8Sepolia = createCommon({
  ...ancient8Sepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var anvil = createCommon({
  ...anvil$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var apexTestnet = createCommon({
  ...apexTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrum = createCommon({
  ...arbitrum$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrumGoerli = createCommon({
  ...arbitrumGoerli$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrumNova = createCommon({
  ...arbitrumNova$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrumSepolia = createCommon({
  ...arbitrumSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var areonNetwork = createCommon({
  ...areonNetwork$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var areonNetworkTestnet = createCommon({
  ...areonNetworkTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var artelaTestnet = createCommon({
  ...artelaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var assetChainTestnet = createCommon({
  ...assetChainTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var astar = createCommon({
  ...astar$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var astarZkEVM = createCommon({
  ...astarZkEVM$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var astarZkyoto = createCommon({
  ...astarZkyoto$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var atletaOlympia = createCommon({
  ...atletaOlympia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var aurora = createCommon({
  ...aurora$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var auroraTestnet = createCommon({
  ...auroraTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var auroria = createCommon({
  ...auroria$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var avalanche = createCommon({
  ...avalanche$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var avalancheFuji = createCommon({
  ...avalancheFuji$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var b3 = createCommon({
  ...b3$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var b3Sepolia = createCommon({
  ...b3Sepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bahamut = createCommon({
  ...bahamut$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var base = createCommon({
  ...base$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var baseGoerli = createCommon({
  ...baseGoerli$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var baseSepolia = createCommon({
  ...baseSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var beam = createCommon({
  ...beam$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var beamTestnet = createCommon({
  ...beamTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bearNetworkChainMainnet = createCommon({
  ...bearNetworkChainMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bearNetworkChainTestnet = createCommon({
  ...bearNetworkChainTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var berachainTestnet = createCommon({
  ...berachainTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var berachainTestnetbArtio = createCommon({
  ...berachainTestnetbArtio$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bevmMainnet = createCommon({
  ...bevmMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitkub = createCommon({
  ...bitkub$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitkubTestnet = createCommon({
  ...bitkubTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitTorrent = createCommon({
  ...bitTorrent$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitTorrentTestnet = createCommon({
  ...bitTorrentTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var blast = createCommon({
  ...blast$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var blastSepolia = createCommon({
  ...blastSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bob = createCommon({
  ...bob$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var boba = createCommon({
  ...boba$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bobaSepolia = createCommon({
  ...bobaSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bobSepolia = createCommon({
  ...bobSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var botanixTestnet = createCommon({
  ...botanixTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bronos = createCommon({
  ...bronos$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bronosTestnet = createCommon({
  ...bronosTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bsc = createCommon({
  ...bsc$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bscGreenfield = createCommon({
  ...bscGreenfield$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bscTestnet = createCommon({
  ...bscTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var btr = createCommon({
  ...btr$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var btrTestnet = createCommon({
  ...btrTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bxn = createCommon({
  ...bxn$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bxnTestnet = createCommon({
  ...bxnTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var canto = createCommon({
  ...canto$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var celo = createCommon({
  ...celo$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var celoAlfajores = createCommon({
  ...celoAlfajores$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var chiliz = createCommon({
  ...chiliz$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var chips = createCommon({
  ...chips$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var classic = createCommon({
  ...classic$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var confluxESpace = createCommon({
  ...confluxESpace$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var confluxESpaceTestnet = createCommon({
  ...confluxESpaceTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var coreDao = createCommon({
  ...coreDao$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var crab = createCommon({
  ...crab$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronos = createCommon({
  ...cronos$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronosTestnet = createCommon({
  ...cronosTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronoszkEVM = createCommon({
  ...cronoszkEVM$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronoszkEVMTestnet = createCommon({
  ...cronoszkEVMTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var crossbell = createCommon({
  ...crossbell$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var curtis = createCommon({
  ...curtis$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cyber = createCommon({
  ...cyber$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cyberTestnet = createCommon({
  ...cyberTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var darwinia = createCommon({
  ...darwinia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dchain = createCommon({
  ...dchain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dchainTestnet = createCommon({
  ...dchainTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var defichainEvm = createCommon({
  ...defichainEvm$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var defichainEvmTestnet = createCommon({
  ...defichainEvmTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var degen = createCommon({
  ...degen$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dfk = createCommon({
  ...dfk$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dodochainTestnet = createCommon({
  ...dodochainTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dogechain = createCommon({
  ...dogechain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dreyerxMainnet = createCommon({
  ...dreyerxMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dreyerxTestnet = createCommon({
  ...dreyerxTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgeless = createCommon({
  ...edgeless$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgelessTestnet = createCommon({
  ...edgelessTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgeware = createCommon({
  ...edgeware$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgewareTestnet = createCommon({
  ...edgewareTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ekta = createCommon({
  ...ekta$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ektaTestnet = createCommon({
  ...ektaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var eon = createCommon({
  ...eon$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var eos = createCommon({
  ...eos$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var eosTestnet = createCommon({
  ...eosTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var etherlink = createCommon({
  ...etherlink$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var etherlinkTestnet = createCommon({
  ...etherlinkTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var evmos = createCommon({
  ...evmos$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var evmosTestnet = createCommon({
  ...evmosTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fantom = createCommon({
  ...fantom$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fantomSonicTestnet = createCommon({
  ...fantomSonicTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fantomTestnet = createCommon({
  ...fantomTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fibo = createCommon({
  ...fibo$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var filecoin = createCommon({
  ...filecoin$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var filecoinCalibration = createCommon({
  ...filecoinCalibration$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var filecoinHyperspace = createCommon({
  ...filecoinHyperspace$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flare = createCommon({
  ...flare$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flareTestnet = createCommon({
  ...flareTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flowMainnet = createCommon({
  ...flowMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flowPreviewnet = createCommon({
  ...flowPreviewnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flowTestnet = createCommon({
  ...flowTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fluence = createCommon({
  ...fluence$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fluenceStage = createCommon({
  ...fluenceStage$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fluenceTestnet = createCommon({
  ...fluenceTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var forma = createCommon({
  ...forma$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var foundry = createCommon({
  ...foundry$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fraxtal = createCommon({
  ...fraxtal$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fraxtalTestnet = createCommon({
  ...fraxtalTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var funkiMainnet = createCommon({
  ...funkiMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var funkiSepolia = createCommon({
  ...funkiSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fuse = createCommon({
  ...fuse$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fuseSparknet = createCommon({
  ...fuseSparknet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gnosis = createCommon({
  ...gnosis$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gnosisChiado = createCommon({
  ...gnosisChiado$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gobi = createCommon({
  ...gobi$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var goerli = createCommon({
  ...goerli$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gravity = createCommon({
  ...gravity$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ham = createCommon({
  ...ham$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var haqqMainnet = createCommon({
  ...haqqMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var haqqTestedge2 = createCommon({
  ...haqqTestedge2$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hardhat = createCommon({
  ...hardhat$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var harmonyOne = createCommon({
  ...harmonyOne$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hashkeyTestnet = createCommon({
  ...hashkeyTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hedera = createCommon({
  ...hedera$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hederaPreviewnet = createCommon({
  ...hederaPreviewnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hederaTestnet = createCommon({
  ...hederaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var holesky = createCommon({
  ...holesky$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var immutableZkEvm = createCommon({
  ...immutableZkEvm$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var immutableZkEvmTestnet = createCommon({
  ...immutableZkEvmTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var inEVM = createCommon({
  ...inEVM$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iota = createCommon({
  ...iota$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iotaTestnet = createCommon({
  ...iotaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iotex = createCommon({
  ...iotex$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iotexTestnet = createCommon({
  ...iotexTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var jbc = createCommon({
  ...jbc$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var jbcTestnet = createCommon({
  ...jbcTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kaia = createCommon({
  ...kaia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kairos = createCommon({
  ...kairos$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kakarotSepolia = createCommon({
  ...kakarotSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var karura = createCommon({
  ...karura$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kava = createCommon({
  ...kava$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kavaTestnet = createCommon({
  ...kavaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kcc = createCommon({
  ...kcc$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var klaytn = createCommon({
  ...klaytn$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var klaytnBaobab = createCommon({
  ...klaytnBaobab$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var koi = createCommon({
  ...koi$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kroma = createCommon({
  ...kroma$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kromaSepolia = createCommon({
  ...kromaSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var l3x = createCommon({
  ...l3x$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var l3xTestnet = createCommon({
  ...l3xTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lightlinkPegasus = createCommon({
  ...lightlinkPegasus$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lightlinkPhoenix = createCommon({
  ...lightlinkPhoenix$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var linea = createCommon({
  ...linea$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lineaGoerli = createCommon({
  ...lineaGoerli$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lineaSepolia = createCommon({
  ...lineaSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lineaTestnet = createCommon({
  ...lineaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lisk = createCommon({
  ...lisk$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var liskSepolia = createCommon({
  ...liskSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var localhost = createCommon({
  ...localhost$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lukso = createCommon({
  ...lukso$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var luksoTestnet = createCommon({
  ...luksoTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lycan = createCommon({
  ...lycan$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lyra = createCommon({
  ...lyra$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mainnet = createCommon({
  ...mainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mandala = createCommon({
  ...mandala$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var manta = createCommon({
  ...manta$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantaSepoliaTestnet = createCommon({
  ...mantaSepoliaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantaTestnet = createCommon({
  ...mantaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantle = createCommon({
  ...mantle$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantleSepoliaTestnet = createCommon({
  ...mantleSepoliaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantleTestnet = createCommon({
  ...mantleTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var merlin = createCommon({
  ...merlin$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metachain = createCommon({
  ...metachain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metachainIstanbul = createCommon({
  ...metachainIstanbul$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metalL2 = createCommon({
  ...metalL2$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var meter = createCommon({
  ...meter$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var meterTestnet = createCommon({
  ...meterTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metis = createCommon({
  ...metis$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metisGoerli = createCommon({
  ...metisGoerli$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mev = createCommon({
  ...mev$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mevTestnet = createCommon({
  ...mevTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mintSepoliaTestnet = createCommon({
  ...mintSepoliaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mode = createCommon({
  ...mode$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var modeTestnet = createCommon({
  ...modeTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonbaseAlpha = createCommon({
  ...moonbaseAlpha$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonbeam = createCommon({
  ...moonbeam$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonbeamDev = createCommon({
  ...moonbeamDev$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonriver = createCommon({
  ...moonriver$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var morphHolesky = createCommon({
  ...morphHolesky$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var morphSepolia = createCommon({
  ...morphSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var nautilus = createCommon({
  ...nautilus$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var neonDevnet = createCommon({
  ...neonDevnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var neonMainnet = createCommon({
  ...neonMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var nexi = createCommon({
  ...nexi$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var nexilix = createCommon({
  ...nexilix$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var oasisTestnet = createCommon({
  ...oasisTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var oasys = createCommon({
  ...oasys$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var okc = createCommon({
  ...okc$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var oortMainnetDev = createCommon({
  ...oortMainnetDev$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var opBNB = createCommon({
  ...opBNB$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var opBNBTestnet = createCommon({
  ...opBNBTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var optimism = createCommon({
  ...optimism$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var optimismGoerli = createCommon({
  ...optimismGoerli$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var optimismSepolia = createCommon({
  ...optimismSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var otimDevnet = createCommon({
  ...otimDevnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var palm = createCommon({
  ...palm$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var palmTestnet = createCommon({
  ...palmTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pgn = createCommon({
  ...pgn$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pgnTestnet = createCommon({
  ...pgnTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var phoenix = createCommon({
  ...phoenix$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var playfiAlbireo = createCommon({
  ...playfiAlbireo$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var plinga = createCommon({
  ...plinga$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var plumeTestnet = createCommon({
  ...plumeTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygon = createCommon({
  ...polygon$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonAmoy = createCommon({
  ...polygonAmoy$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonMumbai = createCommon({
  ...polygonMumbai$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonZkEvm = createCommon({
  ...polygonZkEvm$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonZkEvmCardona = createCommon({
  ...polygonZkEvmCardona$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonZkEvmTestnet = createCommon({
  ...polygonZkEvmTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pulsechain = createCommon({
  ...pulsechain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pulsechainV4 = createCommon({
  ...pulsechainV4$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var qMainnet = createCommon({
  ...qMainnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var qTestnet = createCommon({
  ...qTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var real = createCommon({
  ...real$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var redbellyTestnet = createCommon({
  ...redbellyTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var redstone = createCommon({
  ...redstone$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var reyaNetwork = createCommon({
  ...reyaNetwork$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rollux = createCommon({
  ...rollux$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rolluxTestnet = createCommon({
  ...rolluxTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ronin = createCommon({
  ...ronin$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var root = createCommon({
  ...root$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rootPorcini = createCommon({
  ...rootPorcini$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rootstock = createCommon({
  ...rootstock$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rootstockTestnet = createCommon({
  ...rootstockTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rss3 = createCommon({
  ...rss3$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rss3Sepolia = createCommon({
  ...rss3Sepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var saigon = createCommon({
  ...saigon$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sapphire = createCommon({
  ...sapphire$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sapphireTestnet = createCommon({
  ...sapphireTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var satoshiVM = createCommon({
  ...satoshiVM$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var satoshiVMTestnet = createCommon({
  ...satoshiVMTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var scroll = createCommon({
  ...scroll$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var scrollSepolia = createCommon({
  ...scrollSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sei = createCommon({
  ...sei$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var seiDevnet = createCommon({
  ...seiDevnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var seiTestnet = createCommon({
  ...seiTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sepolia = createCommon({
  ...sepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shapeSepolia = createCommon({
  ...shapeSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shardeumSphinx = createCommon({
  ...shardeumSphinx$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shibarium = createCommon({
  ...shibarium$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shibariumTestnet = createCommon({
  ...shibariumTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shimmer = createCommon({
  ...shimmer$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shimmerTestnet = createCommon({
  ...shimmerTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleBlockBrawlers = createCommon({
  ...skaleBlockBrawlers$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCalypso = createCommon({
  ...skaleCalypso$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCalypsoTestnet = createCommon({
  ...skaleCalypsoTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCryptoBlades = createCommon({
  ...skaleCryptoBlades$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCryptoColosseum = createCommon({
  ...skaleCryptoColosseum$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleEuropa = createCommon({
  ...skaleEuropa$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleEuropaTestnet = createCommon({
  ...skaleEuropaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleExorde = createCommon({
  ...skaleExorde$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleHumanProtocol = createCommon({
  ...skaleHumanProtocol$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleNebula = createCommon({
  ...skaleNebula$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleNebulaTestnet = createCommon({
  ...skaleNebulaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleRazor = createCommon({
  ...skaleRazor$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleTitan = createCommon({
  ...skaleTitan$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleTitanTestnet = createCommon({
  ...skaleTitanTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sketchpad = createCommon({
  ...sketchpad$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var soneiumMinato = createCommon({
  ...soneiumMinato$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var songbird = createCommon({
  ...songbird$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var songbirdTestnet = createCommon({
  ...songbirdTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sophonTestnet = createCommon({
  ...sophonTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var spicy = createCommon({
  ...spicy$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var storyTestnet = createCommon({
  ...storyTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var stratis = createCommon({
  ...stratis$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var syscoin = createCommon({
  ...syscoin$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var syscoinTestnet = createCommon({
  ...syscoinTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taiko = createCommon({
  ...taiko$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoHekla = createCommon({
  ...taikoHekla$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoJolnir = createCommon({
  ...taikoJolnir$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoKatla = createCommon({
  ...taikoKatla$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoTestnetSepolia = createCommon({
  ...taikoTestnetSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taraxa = createCommon({
  ...taraxa$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taraxaTestnet = createCommon({
  ...taraxaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var telcoinTestnet = createCommon({
  ...telcoinTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var telos = createCommon({
  ...telos$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var telosTestnet = createCommon({
  ...telosTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var tenet = createCommon({
  ...tenet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var tevmDefault = createCommon({
  ...defineChain({
    id: 900,
    name: "tevm-devnet",
    fees: optimism$1.fees,
    rpcUrls: foundry$1.rpcUrls,
    testnet: true,
    custom: foundry$1.custom,
    contracts: {
      ...foundry$1.contracts,
      // we add this in createTevmNode
      multicall3: { address: "0xcA11bde05977b3631167028862bE2a173976CA11", blockCreated: 0 }
    },
    formatters: optimism$1.formatters,
    nativeCurrency: optimism$1.nativeCurrency,
    blockExplorers: foundry$1.blockExplorers,
    serializers: optimism$1.serializers
  }),
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var thaiChain = createCommon({
  ...thaiChain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var thunderTestnet = createCommon({
  ...thunderTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var tron = createCommon({
  ...tron$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var unreal = createCommon({
  ...unreal$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var vechain = createCommon({
  ...vechain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wanchain = createCommon({
  ...wanchain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wanchainTestnet = createCommon({
  ...wanchainTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wemix = createCommon({
  ...wemix$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wemixTestnet = createCommon({
  ...wemixTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var x1Testnet = createCommon({
  ...x1Testnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xai = createCommon({
  ...xai$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xaiTestnet = createCommon({
  ...xaiTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xdc = createCommon({
  ...xdc$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xdcTestnet = createCommon({
  ...xdcTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xLayer = createCommon({
  ...xLayer$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xLayerTestnet = createCommon({
  ...xLayerTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xrSepolia = createCommon({
  ...xrSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var yooldoVerse = createCommon({
  ...yooldoVerse$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var yooldoVerseTestnet = createCommon({
  ...yooldoVerseTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zetachain = createCommon({
  ...zetachain$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zetachainAthensTestnet = createCommon({
  ...zetachainAthensTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zhejiang = createCommon({
  ...zhejiang$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zilliqa = createCommon({
  ...zilliqa$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zilliqaTestnet = createCommon({
  ...zilliqaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zircuitTestnet = createCommon({
  ...zircuitTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkFair = createCommon({
  ...zkFair$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkFairTestnet = createCommon({
  ...zkFairTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkLinkNova = createCommon({
  ...zkLinkNova$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkLinkNovaSepoliaTestnet = createCommon({
  ...zkLinkNovaSepoliaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksync = createCommon({
  ...zksync$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksyncInMemoryNode = createCommon({
  ...zksyncInMemoryNode$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksyncLocalNode = createCommon({
  ...zksyncLocalNode$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksyncSepoliaTestnet = createCommon({
  ...zksyncSepoliaTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zora = createCommon({
  ...zora$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zoraSepolia = createCommon({
  ...zoraSepolia$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zoraTestnet = createCommon({
  ...zoraTestnet$1,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});

export { abstractTestnet, acala, ancient8, ancient8Sepolia, anvil, apexTestnet, arbitrum, arbitrumGoerli, arbitrumNova, arbitrumSepolia, areonNetwork, areonNetworkTestnet, artelaTestnet, assetChainTestnet, astar, astarZkEVM, astarZkyoto, atletaOlympia, aurora, auroraTestnet, auroria, avalanche, avalancheFuji, b3, b3Sepolia, bahamut, base, baseGoerli, baseSepolia, beam, beamTestnet, bearNetworkChainMainnet, bearNetworkChainTestnet, berachainTestnet, berachainTestnetbArtio, bevmMainnet, bitTorrent, bitTorrentTestnet, bitkub, bitkubTestnet, blast, blastSepolia, bob, bobSepolia, boba, bobaSepolia, botanixTestnet, bronos, bronosTestnet, bsc, bscGreenfield, bscTestnet, btr, btrTestnet, bxn, bxnTestnet, canto, celo, celoAlfajores, chiliz, chips, classic, confluxESpace, confluxESpaceTestnet, coreDao, crab, createCommon, createMockKzg, cronos, cronosTestnet, cronoszkEVM, cronoszkEVMTestnet, crossbell, curtis, cyber, cyberTestnet, darwinia, dchain, dchainTestnet, defichainEvm, defichainEvmTestnet, degen, dfk, dodochainTestnet, dogechain, dreyerxMainnet, dreyerxTestnet, edgeless, edgelessTestnet, edgeware, edgewareTestnet, ekta, ektaTestnet, eon, eos, eosTestnet, etherlink, etherlinkTestnet, evmos, evmosTestnet, fantom, fantomSonicTestnet, fantomTestnet, fibo, filecoin, filecoinCalibration, filecoinHyperspace, flare, flareTestnet, flowMainnet, flowPreviewnet, flowTestnet, fluence, fluenceStage, fluenceTestnet, forma, foundry, fraxtal, fraxtalTestnet, funkiMainnet, funkiSepolia, fuse, fuseSparknet, gnosis, gnosisChiado, gobi, goerli, gravity, ham, haqqMainnet, haqqTestedge2, hardhat, harmonyOne, hashkeyTestnet, hedera, hederaPreviewnet, hederaTestnet, holesky, immutableZkEvm, immutableZkEvmTestnet, inEVM, iota, iotaTestnet, iotex, iotexTestnet, jbc, jbcTestnet, kaia, kairos, kakarotSepolia, karura, kava, kavaTestnet, kcc, klaytn, klaytnBaobab, koi, kroma, kromaSepolia, l3x, l3xTestnet, lightlinkPegasus, lightlinkPhoenix, linea, lineaGoerli, lineaSepolia, lineaTestnet, lisk, liskSepolia, localhost, lukso, luksoTestnet, lycan, lyra, mainnet, mandala, manta, mantaSepoliaTestnet, mantaTestnet, mantle, mantleSepoliaTestnet, mantleTestnet, merlin, metachain, metachainIstanbul, metalL2, meter, meterTestnet, metis, metisGoerli, mev, mevTestnet, mintSepoliaTestnet, mode, modeTestnet, moonbaseAlpha, moonbeam, moonbeamDev, moonriver, morphHolesky, morphSepolia, nautilus, neonDevnet, neonMainnet, nexi, nexilix, oasisTestnet, oasys, okc, oortMainnetDev, opBNB, opBNBTestnet, optimism, optimismGoerli, optimismSepolia, otimDevnet, palm, palmTestnet, pgn, pgnTestnet, phoenix, playfiAlbireo, plinga, plumeTestnet, polygon, polygonAmoy, polygonMumbai, polygonZkEvm, polygonZkEvmCardona, polygonZkEvmTestnet, pulsechain, pulsechainV4, qMainnet, qTestnet, real, redbellyTestnet, redstone, reyaNetwork, rollux, rolluxTestnet, ronin, root, rootPorcini, rootstock, rootstockTestnet, rss3, rss3Sepolia, saigon, sapphire, sapphireTestnet, satoshiVM, satoshiVMTestnet, scroll, scrollSepolia, sei, seiDevnet, seiTestnet, sepolia, shapeSepolia, shardeumSphinx, shibarium, shibariumTestnet, shimmer, shimmerTestnet, skaleBlockBrawlers, skaleCalypso, skaleCalypsoTestnet, skaleCryptoBlades, skaleCryptoColosseum, skaleEuropa, skaleEuropaTestnet, skaleExorde, skaleHumanProtocol, skaleNebula, skaleNebulaTestnet, skaleRazor, skaleTitan, skaleTitanTestnet, sketchpad, soneiumMinato, songbird, songbirdTestnet, sophonTestnet, spicy, storyTestnet, stratis, syscoin, syscoinTestnet, taiko, taikoHekla, taikoJolnir, taikoKatla, taikoTestnetSepolia, taraxa, taraxaTestnet, telcoinTestnet, telos, telosTestnet, tenet, tevmDefault, thaiChain, thunderTestnet, tron, unreal, vechain, wanchain, wanchainTestnet, wemix, wemixTestnet, x1Testnet, xLayer, xLayerTestnet, xai, xaiTestnet, xdc, xdcTestnet, xrSepolia, yooldoVerse, yooldoVerseTestnet, zetachain, zetachainAthensTestnet, zhejiang, zilliqa, zilliqaTestnet, zircuitTestnet, zkFair, zkFairTestnet, zkLinkNova, zkLinkNovaSepoliaTestnet, zksync, zksyncInMemoryNode, zksyncLocalNode, zksyncSepoliaTestnet, zora, zoraSepolia, zoraTestnet };
