// @tevm/common/dist/index.cjs
'use strict';

var common = require('@ethereumjs/common');
var errors = require('@tevm/errors');
var logger = require('@tevm/logger');
var utils = require('@tevm/utils');
var viem = require('viem');
var chains = require('viem/chains');

// src/index.ts
var createMockKzg = () => {
  const mockHash = viem.bytesToHex(utils.keccak256("0x69", "bytes"));
  return {
    loadTrustedSetup: () => 69,
    verifyKzgProof: () => true,
    freeTrustedSetup: () => {
    },
    verifyBlobKzgProof: () => true,
    blobToKzgCommitment: () => mockHash,
    computeBlobKzgProof: () => mockHash,
    verifyBlobKzgProofBatch: () => true,
    // New methods required by the updated KZG interface
    computeBlobProof: () => mockHash,
    verifyProof: () => true,
    verifyBlobProofBatch: () => true
  };
};

// src/createCommon.js
var createCommon = ({
  customCrypto = {},
  loggingLevel = "warn",
  hardfork = "prague",
  eips = [],
  ...chain
}) => {
  try {
    const logger$1 = logger.createLogger({ level: loggingLevel, name: "@tevm/common" });
    const eipsArray = Array.isArray(eips) ? eips : [];
    const finalCustomCrypto = customCrypto && Object.keys(customCrypto).length > 0 ? { kzg: createMockKzg(), ...customCrypto } : { kzg: createMockKzg() };
    const ethjsCommon = common.createCustomCommon(
      {
        chainId: chain.id,
        name: chain.name || "TevmCustom"
      },
      common.Mainnet,
      {
        hardfork,
        eips: [...eipsArray, 1559, 4895, 4844, 4788, 2935],
        customCrypto: finalCustomCrypto,
        params: {
          1559: {
            elasticityMultiplier: 2,
            baseFeeMaxChangeDenominator: 8,
            initialBaseFee: 1e9
          },
          4844: {
            targetBlobGasPerBlock: 393216,
            blobGasPerBlob: 131072,
            minBlobGasPrice: 1,
            blobGasPriceUpdateFraction: 3338477
          },
          4788: {
            historicalRootsLength: 8191
          },
          // VM params accessed via param('vm', ...)
          vm: {
            historicalRootsLength: 8191,
            historyStorageAddress: "0x0aae40965e6800cd9b1f4b05ff21581047e3f91e",
            historyServeWindow: 8192
          },
          gasConfig: {
            maxRefundQuotient: 5
          }
        }
      }
    );
    if (ethjsCommon.isActivatedEIP(6800)) {
      logger$1.warn("verkle state is currently not supported in tevm");
    }
    logger$1.debug(ethjsCommon.eips(), "Created common with eips enabled");
    return {
      ...chain,
      ethjsCommon,
      copy: () => {
        const ethjsCommonCopy = ethjsCommon.copy();
        const newCommon = createCommon({ loggingLevel, hardfork, eips, ...chain });
        newCommon.ethjsCommon = ethjsCommonCopy;
        return newCommon;
      }
    };
  } catch (e) {
    const err = (
      /** @type {Error} */
      e
    );
    throw new errors.InvalidParamsError(err.message, { cause: err });
  }
};
var abstractTestnet = createCommon({
  ...chains.abstractTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var acala = createCommon({
  ...chains.acala,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ancient8 = createCommon({
  ...chains.ancient8,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ancient8Sepolia = createCommon({
  ...chains.ancient8Sepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var anvil = createCommon({
  ...chains.anvil,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var apexTestnet = createCommon({
  ...chains.apexTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrum = createCommon({
  ...chains.arbitrum,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrumGoerli = createCommon({
  ...chains.arbitrumGoerli,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrumNova = createCommon({
  ...chains.arbitrumNova,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var arbitrumSepolia = createCommon({
  ...chains.arbitrumSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var areonNetwork = createCommon({
  ...chains.areonNetwork,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var areonNetworkTestnet = createCommon({
  ...chains.areonNetworkTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var artelaTestnet = createCommon({
  ...chains.artelaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var assetChainTestnet = createCommon({
  ...chains.assetChainTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var astar = createCommon({
  ...chains.astar,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var astarZkEVM = createCommon({
  ...chains.astarZkEVM,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var astarZkyoto = createCommon({
  ...chains.astarZkyoto,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var atletaOlympia = createCommon({
  ...chains.atletaOlympia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var aurora = createCommon({
  ...chains.aurora,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var auroraTestnet = createCommon({
  ...chains.auroraTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var auroria = createCommon({
  ...chains.auroria,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var avalanche = createCommon({
  ...chains.avalanche,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var avalancheFuji = createCommon({
  ...chains.avalancheFuji,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var b3 = createCommon({
  ...chains.b3,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var b3Sepolia = createCommon({
  ...chains.b3Sepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bahamut = createCommon({
  ...chains.bahamut,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var base = createCommon({
  ...chains.base,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var baseGoerli = createCommon({
  ...chains.baseGoerli,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var baseSepolia = createCommon({
  ...chains.baseSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var beam = createCommon({
  ...chains.beam,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var beamTestnet = createCommon({
  ...chains.beamTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bearNetworkChainMainnet = createCommon({
  ...chains.bearNetworkChainMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bearNetworkChainTestnet = createCommon({
  ...chains.bearNetworkChainTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var berachainTestnet = createCommon({
  ...chains.berachainTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var berachainTestnetbArtio = createCommon({
  ...chains.berachainTestnetbArtio,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bevmMainnet = createCommon({
  ...chains.bevmMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitkub = createCommon({
  ...chains.bitkub,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitkubTestnet = createCommon({
  ...chains.bitkubTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitTorrent = createCommon({
  ...chains.bitTorrent,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bitTorrentTestnet = createCommon({
  ...chains.bitTorrentTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var blast = createCommon({
  ...chains.blast,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var blastSepolia = createCommon({
  ...chains.blastSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bob = createCommon({
  ...chains.bob,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var boba = createCommon({
  ...chains.boba,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bobaSepolia = createCommon({
  ...chains.bobaSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bobSepolia = createCommon({
  ...chains.bobSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var botanixTestnet = createCommon({
  ...chains.botanixTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bronos = createCommon({
  ...chains.bronos,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bronosTestnet = createCommon({
  ...chains.bronosTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bsc = createCommon({
  ...chains.bsc,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bscGreenfield = createCommon({
  ...chains.bscGreenfield,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bscTestnet = createCommon({
  ...chains.bscTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var btr = createCommon({
  ...chains.btr,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var btrTestnet = createCommon({
  ...chains.btrTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bxn = createCommon({
  ...chains.bxn,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var bxnTestnet = createCommon({
  ...chains.bxnTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var canto = createCommon({
  ...chains.canto,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var celo = createCommon({
  ...chains.celo,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var celoAlfajores = createCommon({
  ...chains.celoAlfajores,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var chiliz = createCommon({
  ...chains.chiliz,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var chips = createCommon({
  ...chains.chips,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var classic = createCommon({
  ...chains.classic,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var confluxESpace = createCommon({
  ...chains.confluxESpace,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var confluxESpaceTestnet = createCommon({
  ...chains.confluxESpaceTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var coreDao = createCommon({
  ...chains.coreDao,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var crab = createCommon({
  ...chains.crab,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronos = createCommon({
  ...chains.cronos,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronosTestnet = createCommon({
  ...chains.cronosTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronoszkEVM = createCommon({
  ...chains.cronoszkEVM,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cronoszkEVMTestnet = createCommon({
  ...chains.cronoszkEVMTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var crossbell = createCommon({
  ...chains.crossbell,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var curtis = createCommon({
  ...chains.curtis,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cyber = createCommon({
  ...chains.cyber,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var cyberTestnet = createCommon({
  ...chains.cyberTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var darwinia = createCommon({
  ...chains.darwinia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dchain = createCommon({
  ...chains.dchain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dchainTestnet = createCommon({
  ...chains.dchainTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var defichainEvm = createCommon({
  ...chains.defichainEvm,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var defichainEvmTestnet = createCommon({
  ...chains.defichainEvmTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var degen = createCommon({
  ...chains.degen,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dfk = createCommon({
  ...chains.dfk,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dodochainTestnet = createCommon({
  ...chains.dodochainTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dogechain = createCommon({
  ...chains.dogechain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dreyerxMainnet = createCommon({
  ...chains.dreyerxMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var dreyerxTestnet = createCommon({
  ...chains.dreyerxTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgeless = createCommon({
  ...chains.edgeless,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgelessTestnet = createCommon({
  ...chains.edgelessTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgeware = createCommon({
  ...chains.edgeware,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var edgewareTestnet = createCommon({
  ...chains.edgewareTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ekta = createCommon({
  ...chains.ekta,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ektaTestnet = createCommon({
  ...chains.ektaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var eon = createCommon({
  ...chains.eon,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var eos = createCommon({
  ...chains.eos,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var eosTestnet = createCommon({
  ...chains.eosTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var etherlink = createCommon({
  ...chains.etherlink,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var etherlinkTestnet = createCommon({
  ...chains.etherlinkTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var evmos = createCommon({
  ...chains.evmos,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var evmosTestnet = createCommon({
  ...chains.evmosTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fantom = createCommon({
  ...chains.fantom,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fantomSonicTestnet = createCommon({
  ...chains.fantomSonicTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fantomTestnet = createCommon({
  ...chains.fantomTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fibo = createCommon({
  ...chains.fibo,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var filecoin = createCommon({
  ...chains.filecoin,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var filecoinCalibration = createCommon({
  ...chains.filecoinCalibration,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var filecoinHyperspace = createCommon({
  ...chains.filecoinHyperspace,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flare = createCommon({
  ...chains.flare,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flareTestnet = createCommon({
  ...chains.flareTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flowMainnet = createCommon({
  ...chains.flowMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flowPreviewnet = createCommon({
  ...chains.flowPreviewnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var flowTestnet = createCommon({
  ...chains.flowTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fluence = createCommon({
  ...chains.fluence,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fluenceStage = createCommon({
  ...chains.fluenceStage,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fluenceTestnet = createCommon({
  ...chains.fluenceTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var forma = createCommon({
  ...chains.forma,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var foundry = createCommon({
  ...chains.foundry,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fraxtal = createCommon({
  ...chains.fraxtal,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fraxtalTestnet = createCommon({
  ...chains.fraxtalTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var funkiMainnet = createCommon({
  ...chains.funkiMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var funkiSepolia = createCommon({
  ...chains.funkiSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fuse = createCommon({
  ...chains.fuse,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var fuseSparknet = createCommon({
  ...chains.fuseSparknet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gnosis = createCommon({
  ...chains.gnosis,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gnosisChiado = createCommon({
  ...chains.gnosisChiado,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gobi = createCommon({
  ...chains.gobi,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var goerli = createCommon({
  ...chains.goerli,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var gravity = createCommon({
  ...chains.gravity,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ham = createCommon({
  ...chains.ham,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var haqqMainnet = createCommon({
  ...chains.haqqMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var haqqTestedge2 = createCommon({
  ...chains.haqqTestedge2,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hardhat = createCommon({
  ...chains.hardhat,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var harmonyOne = createCommon({
  ...chains.harmonyOne,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hashkeyTestnet = createCommon({
  ...chains.hashkeyTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hedera = createCommon({
  ...chains.hedera,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hederaPreviewnet = createCommon({
  ...chains.hederaPreviewnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var hederaTestnet = createCommon({
  ...chains.hederaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var holesky = createCommon({
  ...chains.holesky,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var immutableZkEvm = createCommon({
  ...chains.immutableZkEvm,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var immutableZkEvmTestnet = createCommon({
  ...chains.immutableZkEvmTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var inEVM = createCommon({
  ...chains.inEVM,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iota = createCommon({
  ...chains.iota,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iotaTestnet = createCommon({
  ...chains.iotaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iotex = createCommon({
  ...chains.iotex,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var iotexTestnet = createCommon({
  ...chains.iotexTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var jbc = createCommon({
  ...chains.jbc,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var jbcTestnet = createCommon({
  ...chains.jbcTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kaia = createCommon({
  ...chains.kaia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kairos = createCommon({
  ...chains.kairos,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kakarotSepolia = createCommon({
  ...chains.kakarotSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var karura = createCommon({
  ...chains.karura,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kava = createCommon({
  ...chains.kava,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kavaTestnet = createCommon({
  ...chains.kavaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kcc = createCommon({
  ...chains.kcc,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var klaytn = createCommon({
  ...chains.klaytn,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var klaytnBaobab = createCommon({
  ...chains.klaytnBaobab,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var koi = createCommon({
  ...chains.koi,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kroma = createCommon({
  ...chains.kroma,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var kromaSepolia = createCommon({
  ...chains.kromaSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var l3x = createCommon({
  ...chains.l3x,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var l3xTestnet = createCommon({
  ...chains.l3xTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lightlinkPegasus = createCommon({
  ...chains.lightlinkPegasus,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lightlinkPhoenix = createCommon({
  ...chains.lightlinkPhoenix,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var linea = createCommon({
  ...chains.linea,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lineaGoerli = createCommon({
  ...chains.lineaGoerli,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lineaSepolia = createCommon({
  ...chains.lineaSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lineaTestnet = createCommon({
  ...chains.lineaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lisk = createCommon({
  ...chains.lisk,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var liskSepolia = createCommon({
  ...chains.liskSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var localhost = createCommon({
  ...chains.localhost,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lukso = createCommon({
  ...chains.lukso,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var luksoTestnet = createCommon({
  ...chains.luksoTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lycan = createCommon({
  ...chains.lycan,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var lyra = createCommon({
  ...chains.lyra,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mainnet = createCommon({
  ...chains.mainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mandala = createCommon({
  ...chains.mandala,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var manta = createCommon({
  ...chains.manta,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantaSepoliaTestnet = createCommon({
  ...chains.mantaSepoliaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantaTestnet = createCommon({
  ...chains.mantaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantle = createCommon({
  ...chains.mantle,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantleSepoliaTestnet = createCommon({
  ...chains.mantleSepoliaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mantleTestnet = createCommon({
  ...chains.mantleTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var merlin = createCommon({
  ...chains.merlin,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metachain = createCommon({
  ...chains.metachain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metachainIstanbul = createCommon({
  ...chains.metachainIstanbul,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metalL2 = createCommon({
  ...chains.metalL2,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var meter = createCommon({
  ...chains.meter,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var meterTestnet = createCommon({
  ...chains.meterTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metis = createCommon({
  ...chains.metis,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var metisGoerli = createCommon({
  ...chains.metisGoerli,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mev = createCommon({
  ...chains.mev,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mevTestnet = createCommon({
  ...chains.mevTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mintSepoliaTestnet = createCommon({
  ...chains.mintSepoliaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var mode = createCommon({
  ...chains.mode,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var modeTestnet = createCommon({
  ...chains.modeTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonbaseAlpha = createCommon({
  ...chains.moonbaseAlpha,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonbeam = createCommon({
  ...chains.moonbeam,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonbeamDev = createCommon({
  ...chains.moonbeamDev,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var moonriver = createCommon({
  ...chains.moonriver,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var morphHolesky = createCommon({
  ...chains.morphHolesky,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var morphSepolia = createCommon({
  ...chains.morphSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var nautilus = createCommon({
  ...chains.nautilus,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var neonDevnet = createCommon({
  ...chains.neonDevnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var neonMainnet = createCommon({
  ...chains.neonMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var nexi = createCommon({
  ...chains.nexi,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var nexilix = createCommon({
  ...chains.nexilix,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var oasisTestnet = createCommon({
  ...chains.oasisTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var oasys = createCommon({
  ...chains.oasys,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var okc = createCommon({
  ...chains.okc,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var oortMainnetDev = createCommon({
  ...chains.oortMainnetDev,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var opBNB = createCommon({
  ...chains.opBNB,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var opBNBTestnet = createCommon({
  ...chains.opBNBTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var optimism = createCommon({
  ...chains.optimism,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var optimismGoerli = createCommon({
  ...chains.optimismGoerli,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var optimismSepolia = createCommon({
  ...chains.optimismSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var otimDevnet = createCommon({
  ...chains.otimDevnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var palm = createCommon({
  ...chains.palm,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var palmTestnet = createCommon({
  ...chains.palmTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pgn = createCommon({
  ...chains.pgn,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pgnTestnet = createCommon({
  ...chains.pgnTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var phoenix = createCommon({
  ...chains.phoenix,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var playfiAlbireo = createCommon({
  ...chains.playfiAlbireo,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var plinga = createCommon({
  ...chains.plinga,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var plumeTestnet = createCommon({
  ...chains.plumeTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygon = createCommon({
  ...chains.polygon,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonAmoy = createCommon({
  ...chains.polygonAmoy,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonMumbai = createCommon({
  ...chains.polygonMumbai,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonZkEvm = createCommon({
  ...chains.polygonZkEvm,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonZkEvmCardona = createCommon({
  ...chains.polygonZkEvmCardona,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var polygonZkEvmTestnet = createCommon({
  ...chains.polygonZkEvmTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pulsechain = createCommon({
  ...chains.pulsechain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var pulsechainV4 = createCommon({
  ...chains.pulsechainV4,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var qMainnet = createCommon({
  ...chains.qMainnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var qTestnet = createCommon({
  ...chains.qTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var real = createCommon({
  ...chains.real,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var redbellyTestnet = createCommon({
  ...chains.redbellyTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var redstone = createCommon({
  ...chains.redstone,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var reyaNetwork = createCommon({
  ...chains.reyaNetwork,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rollux = createCommon({
  ...chains.rollux,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rolluxTestnet = createCommon({
  ...chains.rolluxTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var ronin = createCommon({
  ...chains.ronin,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var root = createCommon({
  ...chains.root,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rootPorcini = createCommon({
  ...chains.rootPorcini,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rootstock = createCommon({
  ...chains.rootstock,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rootstockTestnet = createCommon({
  ...chains.rootstockTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rss3 = createCommon({
  ...chains.rss3,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var rss3Sepolia = createCommon({
  ...chains.rss3Sepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var saigon = createCommon({
  ...chains.saigon,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sapphire = createCommon({
  ...chains.sapphire,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sapphireTestnet = createCommon({
  ...chains.sapphireTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var satoshiVM = createCommon({
  ...chains.satoshiVM,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var satoshiVMTestnet = createCommon({
  ...chains.satoshiVMTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var scroll = createCommon({
  ...chains.scroll,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var scrollSepolia = createCommon({
  ...chains.scrollSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sei = createCommon({
  ...chains.sei,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var seiDevnet = createCommon({
  ...chains.seiDevnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var seiTestnet = createCommon({
  ...chains.seiTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sepolia = createCommon({
  ...chains.sepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shapeSepolia = createCommon({
  ...chains.shapeSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shardeumSphinx = createCommon({
  ...chains.shardeumSphinx,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shibarium = createCommon({
  ...chains.shibarium,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shibariumTestnet = createCommon({
  ...chains.shibariumTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shimmer = createCommon({
  ...chains.shimmer,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var shimmerTestnet = createCommon({
  ...chains.shimmerTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleBlockBrawlers = createCommon({
  ...chains.skaleBlockBrawlers,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCalypso = createCommon({
  ...chains.skaleCalypso,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCalypsoTestnet = createCommon({
  ...chains.skaleCalypsoTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCryptoBlades = createCommon({
  ...chains.skaleCryptoBlades,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleCryptoColosseum = createCommon({
  ...chains.skaleCryptoColosseum,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleEuropa = createCommon({
  ...chains.skaleEuropa,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleEuropaTestnet = createCommon({
  ...chains.skaleEuropaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleExorde = createCommon({
  ...chains.skaleExorde,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleHumanProtocol = createCommon({
  ...chains.skaleHumanProtocol,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleNebula = createCommon({
  ...chains.skaleNebula,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleNebulaTestnet = createCommon({
  ...chains.skaleNebulaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleRazor = createCommon({
  ...chains.skaleRazor,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleTitan = createCommon({
  ...chains.skaleTitan,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var skaleTitanTestnet = createCommon({
  ...chains.skaleTitanTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sketchpad = createCommon({
  ...chains.sketchpad,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var soneiumMinato = createCommon({
  ...chains.soneiumMinato,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var songbird = createCommon({
  ...chains.songbird,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var songbirdTestnet = createCommon({
  ...chains.songbirdTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var sophonTestnet = createCommon({
  ...chains.sophonTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var spicy = createCommon({
  ...chains.spicy,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var storyTestnet = createCommon({
  ...chains.storyTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var stratis = createCommon({
  ...chains.stratis,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var syscoin = createCommon({
  ...chains.syscoin,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var syscoinTestnet = createCommon({
  ...chains.syscoinTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taiko = createCommon({
  ...chains.taiko,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoHekla = createCommon({
  ...chains.taikoHekla,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoJolnir = createCommon({
  ...chains.taikoJolnir,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoKatla = createCommon({
  ...chains.taikoKatla,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taikoTestnetSepolia = createCommon({
  ...chains.taikoTestnetSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taraxa = createCommon({
  ...chains.taraxa,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var taraxaTestnet = createCommon({
  ...chains.taraxaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var telcoinTestnet = createCommon({
  ...chains.telcoinTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var telos = createCommon({
  ...chains.telos,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var telosTestnet = createCommon({
  ...chains.telosTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var tenet = createCommon({
  ...chains.tenet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var tevmDefault = createCommon({
  ...viem.defineChain({
    id: 900,
    name: "tevm-devnet",
    fees: chains.optimism.fees,
    rpcUrls: chains.foundry.rpcUrls,
    testnet: true,
    custom: chains.foundry.custom,
    contracts: {
      ...chains.foundry.contracts,
      // we add this in createTevmNode
      multicall3: { address: "0xcA11bde05977b3631167028862bE2a173976CA11", blockCreated: 0 }
    },
    formatters: chains.optimism.formatters,
    nativeCurrency: chains.optimism.nativeCurrency,
    blockExplorers: chains.foundry.blockExplorers,
    serializers: chains.optimism.serializers
  }),
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var thaiChain = createCommon({
  ...chains.thaiChain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var thunderTestnet = createCommon({
  ...chains.thunderTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var tron = createCommon({
  ...chains.tron,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var unreal = createCommon({
  ...chains.unreal,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var vechain = createCommon({
  ...chains.vechain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wanchain = createCommon({
  ...chains.wanchain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wanchainTestnet = createCommon({
  ...chains.wanchainTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wemix = createCommon({
  ...chains.wemix,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var wemixTestnet = createCommon({
  ...chains.wemixTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var x1Testnet = createCommon({
  ...chains.x1Testnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xai = createCommon({
  ...chains.xai,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xaiTestnet = createCommon({
  ...chains.xaiTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xdc = createCommon({
  ...chains.xdc,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xdcTestnet = createCommon({
  ...chains.xdcTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xLayer = createCommon({
  ...chains.xLayer,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xLayerTestnet = createCommon({
  ...chains.xLayerTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var xrSepolia = createCommon({
  ...chains.xrSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var yooldoVerse = createCommon({
  ...chains.yooldoVerse,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var yooldoVerseTestnet = createCommon({
  ...chains.yooldoVerseTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zetachain = createCommon({
  ...chains.zetachain,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zetachainAthensTestnet = createCommon({
  ...chains.zetachainAthensTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zhejiang = createCommon({
  ...chains.zhejiang,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zilliqa = createCommon({
  ...chains.zilliqa,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zilliqaTestnet = createCommon({
  ...chains.zilliqaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zircuitTestnet = createCommon({
  ...chains.zircuitTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkFair = createCommon({
  ...chains.zkFair,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkFairTestnet = createCommon({
  ...chains.zkFairTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkLinkNova = createCommon({
  ...chains.zkLinkNova,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zkLinkNovaSepoliaTestnet = createCommon({
  ...chains.zkLinkNovaSepoliaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksync = createCommon({
  ...chains.zksync,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksyncInMemoryNode = createCommon({
  ...chains.zksyncInMemoryNode,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksyncLocalNode = createCommon({
  ...chains.zksyncLocalNode,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zksyncSepoliaTestnet = createCommon({
  ...chains.zksyncSepoliaTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zora = createCommon({
  ...chains.zora,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zoraSepolia = createCommon({
  ...chains.zoraSepolia,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});
var zoraTestnet = createCommon({
  ...chains.zoraTestnet,
  loggingLevel: "warn",
  eips: [],
  hardfork: "prague"
});

Object.defineProperty(exports, "ConsensusAlgorithm", {
  enumerable: true,
  get: function () { return common.ConsensusAlgorithm; }
});
Object.defineProperty(exports, "ConsensusType", {
  enumerable: true,
  get: function () { return common.ConsensusType; }
});
exports.abstractTestnet = abstractTestnet;
exports.acala = acala;
exports.ancient8 = ancient8;
exports.ancient8Sepolia = ancient8Sepolia;
exports.anvil = anvil;
exports.apexTestnet = apexTestnet;
exports.arbitrum = arbitrum;
exports.arbitrumGoerli = arbitrumGoerli;
exports.arbitrumNova = arbitrumNova;
exports.arbitrumSepolia = arbitrumSepolia;
exports.areonNetwork = areonNetwork;
exports.areonNetworkTestnet = areonNetworkTestnet;
exports.artelaTestnet = artelaTestnet;
exports.assetChainTestnet = assetChainTestnet;
exports.astar = astar;
exports.astarZkEVM = astarZkEVM;
exports.astarZkyoto = astarZkyoto;
exports.atletaOlympia = atletaOlympia;
exports.aurora = aurora;
exports.auroraTestnet = auroraTestnet;
exports.auroria = auroria;
exports.avalanche = avalanche;
exports.avalancheFuji = avalancheFuji;
exports.b3 = b3;
exports.b3Sepolia = b3Sepolia;
exports.bahamut = bahamut;
exports.base = base;
exports.baseGoerli = baseGoerli;
exports.baseSepolia = baseSepolia;
exports.beam = beam;
exports.beamTestnet = beamTestnet;
exports.bearNetworkChainMainnet = bearNetworkChainMainnet;
exports.bearNetworkChainTestnet = bearNetworkChainTestnet;
exports.berachainTestnet = berachainTestnet;
exports.berachainTestnetbArtio = berachainTestnetbArtio;
exports.bevmMainnet = bevmMainnet;
exports.bitTorrent = bitTorrent;
exports.bitTorrentTestnet = bitTorrentTestnet;
exports.bitkub = bitkub;
exports.bitkubTestnet = bitkubTestnet;
exports.blast = blast;
exports.blastSepolia = blastSepolia;
exports.bob = bob;
exports.bobSepolia = bobSepolia;
exports.boba = boba;
exports.bobaSepolia = bobaSepolia;
exports.botanixTestnet = botanixTestnet;
exports.bronos = bronos;
exports.bronosTestnet = bronosTestnet;
exports.bsc = bsc;
exports.bscGreenfield = bscGreenfield;
exports.bscTestnet = bscTestnet;
exports.btr = btr;
exports.btrTestnet = btrTestnet;
exports.bxn = bxn;
exports.bxnTestnet = bxnTestnet;
exports.canto = canto;
exports.celo = celo;
exports.celoAlfajores = celoAlfajores;
exports.chiliz = chiliz;
exports.chips = chips;
exports.classic = classic;
exports.confluxESpace = confluxESpace;
exports.confluxESpaceTestnet = confluxESpaceTestnet;
exports.coreDao = coreDao;
exports.crab = crab;
exports.createCommon = createCommon;
exports.createMockKzg = createMockKzg;
exports.cronos = cronos;
exports.cronosTestnet = cronosTestnet;
exports.cronoszkEVM = cronoszkEVM;
exports.cronoszkEVMTestnet = cronoszkEVMTestnet;
exports.crossbell = crossbell;
exports.curtis = curtis;
exports.cyber = cyber;
exports.cyberTestnet = cyberTestnet;
exports.darwinia = darwinia;
exports.dchain = dchain;
exports.dchainTestnet = dchainTestnet;
exports.defichainEvm = defichainEvm;
exports.defichainEvmTestnet = defichainEvmTestnet;
exports.degen = degen;
exports.dfk = dfk;
exports.dodochainTestnet = dodochainTestnet;
exports.dogechain = dogechain;
exports.dreyerxMainnet = dreyerxMainnet;
exports.dreyerxTestnet = dreyerxTestnet;
exports.edgeless = edgeless;
exports.edgelessTestnet = edgelessTestnet;
exports.edgeware = edgeware;
exports.edgewareTestnet = edgewareTestnet;
exports.ekta = ekta;
exports.ektaTestnet = ektaTestnet;
exports.eon = eon;
exports.eos = eos;
exports.eosTestnet = eosTestnet;
exports.etherlink = etherlink;
exports.etherlinkTestnet = etherlinkTestnet;
exports.evmos = evmos;
exports.evmosTestnet = evmosTestnet;
exports.fantom = fantom;
exports.fantomSonicTestnet = fantomSonicTestnet;
exports.fantomTestnet = fantomTestnet;
exports.fibo = fibo;
exports.filecoin = filecoin;
exports.filecoinCalibration = filecoinCalibration;
exports.filecoinHyperspace = filecoinHyperspace;
exports.flare = flare;
exports.flareTestnet = flareTestnet;
exports.flowMainnet = flowMainnet;
exports.flowPreviewnet = flowPreviewnet;
exports.flowTestnet = flowTestnet;
exports.fluence = fluence;
exports.fluenceStage = fluenceStage;
exports.fluenceTestnet = fluenceTestnet;
exports.forma = forma;
exports.foundry = foundry;
exports.fraxtal = fraxtal;
exports.fraxtalTestnet = fraxtalTestnet;
exports.funkiMainnet = funkiMainnet;
exports.funkiSepolia = funkiSepolia;
exports.fuse = fuse;
exports.fuseSparknet = fuseSparknet;
exports.gnosis = gnosis;
exports.gnosisChiado = gnosisChiado;
exports.gobi = gobi;
exports.goerli = goerli;
exports.gravity = gravity;
exports.ham = ham;
exports.haqqMainnet = haqqMainnet;
exports.haqqTestedge2 = haqqTestedge2;
exports.hardhat = hardhat;
exports.harmonyOne = harmonyOne;
exports.hashkeyTestnet = hashkeyTestnet;
exports.hedera = hedera;
exports.hederaPreviewnet = hederaPreviewnet;
exports.hederaTestnet = hederaTestnet;
exports.holesky = holesky;
exports.immutableZkEvm = immutableZkEvm;
exports.immutableZkEvmTestnet = immutableZkEvmTestnet;
exports.inEVM = inEVM;
exports.iota = iota;
exports.iotaTestnet = iotaTestnet;
exports.iotex = iotex;
exports.iotexTestnet = iotexTestnet;
exports.jbc = jbc;
exports.jbcTestnet = jbcTestnet;
exports.kaia = kaia;
exports.kairos = kairos;
exports.kakarotSepolia = kakarotSepolia;
exports.karura = karura;
exports.kava = kava;
exports.kavaTestnet = kavaTestnet;
exports.kcc = kcc;
exports.klaytn = klaytn;
exports.klaytnBaobab = klaytnBaobab;
exports.koi = koi;
exports.kroma = kroma;
exports.kromaSepolia = kromaSepolia;
exports.l3x = l3x;
exports.l3xTestnet = l3xTestnet;
exports.lightlinkPegasus = lightlinkPegasus;
exports.lightlinkPhoenix = lightlinkPhoenix;
exports.linea = linea;
exports.lineaGoerli = lineaGoerli;
exports.lineaSepolia = lineaSepolia;
exports.lineaTestnet = lineaTestnet;
exports.lisk = lisk;
exports.liskSepolia = liskSepolia;
exports.localhost = localhost;
exports.lukso = lukso;
exports.luksoTestnet = luksoTestnet;
exports.lycan = lycan;
exports.lyra = lyra;
exports.mainnet = mainnet;
exports.mandala = mandala;
exports.manta = manta;
exports.mantaSepoliaTestnet = mantaSepoliaTestnet;
exports.mantaTestnet = mantaTestnet;
exports.mantle = mantle;
exports.mantleSepoliaTestnet = mantleSepoliaTestnet;
exports.mantleTestnet = mantleTestnet;
exports.merlin = merlin;
exports.metachain = metachain;
exports.metachainIstanbul = metachainIstanbul;
exports.metalL2 = metalL2;
exports.meter = meter;
exports.meterTestnet = meterTestnet;
exports.metis = metis;
exports.metisGoerli = metisGoerli;
exports.mev = mev;
exports.mevTestnet = mevTestnet;
exports.mintSepoliaTestnet = mintSepoliaTestnet;
exports.mode = mode;
exports.modeTestnet = modeTestnet;
exports.moonbaseAlpha = moonbaseAlpha;
exports.moonbeam = moonbeam;
exports.moonbeamDev = moonbeamDev;
exports.moonriver = moonriver;
exports.morphHolesky = morphHolesky;
exports.morphSepolia = morphSepolia;
exports.nautilus = nautilus;
exports.neonDevnet = neonDevnet;
exports.neonMainnet = neonMainnet;
exports.nexi = nexi;
exports.nexilix = nexilix;
exports.oasisTestnet = oasisTestnet;
exports.oasys = oasys;
exports.okc = okc;
exports.oortMainnetDev = oortMainnetDev;
exports.opBNB = opBNB;
exports.opBNBTestnet = opBNBTestnet;
exports.optimism = optimism;
exports.optimismGoerli = optimismGoerli;
exports.optimismSepolia = optimismSepolia;
exports.otimDevnet = otimDevnet;
exports.palm = palm;
exports.palmTestnet = palmTestnet;
exports.pgn = pgn;
exports.pgnTestnet = pgnTestnet;
exports.phoenix = phoenix;
exports.playfiAlbireo = playfiAlbireo;
exports.plinga = plinga;
exports.plumeTestnet = plumeTestnet;
exports.polygon = polygon;
exports.polygonAmoy = polygonAmoy;
exports.polygonMumbai = polygonMumbai;
exports.polygonZkEvm = polygonZkEvm;
exports.polygonZkEvmCardona = polygonZkEvmCardona;
exports.polygonZkEvmTestnet = polygonZkEvmTestnet;
exports.pulsechain = pulsechain;
exports.pulsechainV4 = pulsechainV4;
exports.qMainnet = qMainnet;
exports.qTestnet = qTestnet;
exports.real = real;
exports.redbellyTestnet = redbellyTestnet;
exports.redstone = redstone;
exports.reyaNetwork = reyaNetwork;
exports.rollux = rollux;
exports.rolluxTestnet = rolluxTestnet;
exports.ronin = ronin;
exports.root = root;
exports.rootPorcini = rootPorcini;
exports.rootstock = rootstock;
exports.rootstockTestnet = rootstockTestnet;
exports.rss3 = rss3;
exports.rss3Sepolia = rss3Sepolia;
exports.saigon = saigon;
exports.sapphire = sapphire;
exports.sapphireTestnet = sapphireTestnet;
exports.satoshiVM = satoshiVM;
exports.satoshiVMTestnet = satoshiVMTestnet;
exports.scroll = scroll;
exports.scrollSepolia = scrollSepolia;
exports.sei = sei;
exports.seiDevnet = seiDevnet;
exports.seiTestnet = seiTestnet;
exports.sepolia = sepolia;
exports.shapeSepolia = shapeSepolia;
exports.shardeumSphinx = shardeumSphinx;
exports.shibarium = shibarium;
exports.shibariumTestnet = shibariumTestnet;
exports.shimmer = shimmer;
exports.shimmerTestnet = shimmerTestnet;
exports.skaleBlockBrawlers = skaleBlockBrawlers;
exports.skaleCalypso = skaleCalypso;
exports.skaleCalypsoTestnet = skaleCalypsoTestnet;
exports.skaleCryptoBlades = skaleCryptoBlades;
exports.skaleCryptoColosseum = skaleCryptoColosseum;
exports.skaleEuropa = skaleEuropa;
exports.skaleEuropaTestnet = skaleEuropaTestnet;
exports.skaleExorde = skaleExorde;
exports.skaleHumanProtocol = skaleHumanProtocol;
exports.skaleNebula = skaleNebula;
exports.skaleNebulaTestnet = skaleNebulaTestnet;
exports.skaleRazor = skaleRazor;
exports.skaleTitan = skaleTitan;
exports.skaleTitanTestnet = skaleTitanTestnet;
exports.sketchpad = sketchpad;
exports.soneiumMinato = soneiumMinato;
exports.songbird = songbird;
exports.songbirdTestnet = songbirdTestnet;
exports.sophonTestnet = sophonTestnet;
exports.spicy = spicy;
exports.storyTestnet = storyTestnet;
exports.stratis = stratis;
exports.syscoin = syscoin;
exports.syscoinTestnet = syscoinTestnet;
exports.taiko = taiko;
exports.taikoHekla = taikoHekla;
exports.taikoJolnir = taikoJolnir;
exports.taikoKatla = taikoKatla;
exports.taikoTestnetSepolia = taikoTestnetSepolia;
exports.taraxa = taraxa;
exports.taraxaTestnet = taraxaTestnet;
exports.telcoinTestnet = telcoinTestnet;
exports.telos = telos;
exports.telosTestnet = telosTestnet;
exports.tenet = tenet;
exports.tevmDefault = tevmDefault;
exports.thaiChain = thaiChain;
exports.thunderTestnet = thunderTestnet;
exports.tron = tron;
exports.unreal = unreal;
exports.vechain = vechain;
exports.wanchain = wanchain;
exports.wanchainTestnet = wanchainTestnet;
exports.wemix = wemix;
exports.wemixTestnet = wemixTestnet;
exports.x1Testnet = x1Testnet;
exports.xLayer = xLayer;
exports.xLayerTestnet = xLayerTestnet;
exports.xai = xai;
exports.xaiTestnet = xaiTestnet;
exports.xdc = xdc;
exports.xdcTestnet = xdcTestnet;
exports.xrSepolia = xrSepolia;
exports.yooldoVerse = yooldoVerse;
exports.yooldoVerseTestnet = yooldoVerseTestnet;
exports.zetachain = zetachain;
exports.zetachainAthensTestnet = zetachainAthensTestnet;
exports.zhejiang = zhejiang;
exports.zilliqa = zilliqa;
exports.zilliqaTestnet = zilliqaTestnet;
exports.zircuitTestnet = zircuitTestnet;
exports.zkFair = zkFair;
exports.zkFairTestnet = zkFairTestnet;
exports.zkLinkNova = zkLinkNova;
exports.zkLinkNovaSepoliaTestnet = zkLinkNovaSepoliaTestnet;
exports.zksync = zksync;
exports.zksyncInMemoryNode = zksyncInMemoryNode;
exports.zksyncLocalNode = zksyncLocalNode;
exports.zksyncSepoliaTestnet = zksyncSepoliaTestnet;
exports.zora = zora;
exports.zoraSepolia = zoraSepolia;
exports.zoraTestnet = zoraTestnet;
