// @tevm/errors/dist/index.cjs
'use strict';

var evm = require('@ethereumjs/evm');
var viem = require('viem');

// src/ethereum/BaseError.js
var getVersion = () => "1.1.0.next-73";
var BaseError = class _BaseError extends Error {
  /**
   * @param {string} shortMessage - A short, human-readable summary of the error.
   * @param {BaseErrorParameters} args={} - Additional parameters for the error.
   * @param {string} _tag - Internal tag for the error.
   * @param {number} [code] - Error code analogous to the code in JSON RPC error.
   */
  constructor(shortMessage, args, _tag, code = 0) {
    if (new.target === _BaseError) {
      throw new TypeError("Cannot construct BaseError instances directly");
    }
    super();
    const details = (() => {
      if (args.cause === null || args.cause === void 0) {
        return "";
      }
      if (args.cause instanceof _BaseError) {
        return args.cause.docsPath;
      }
      if (typeof args.cause === "string") {
        return args.cause;
      }
      if (typeof args.cause !== "object") {
        return args;
      }
      if ("message" in args.cause) {
        return (
          /** @type {string}*/
          args.cause.message
        );
      }
      if (args.cause instanceof Error) {
        if ("errorType" in args.cause) {
          return args.cause.errorType;
        }
      }
      try {
        return JSON.stringify(args.cause);
      } catch (_e) {
        return "Unable to parse error details";
      }
    })();
    const docsPath = args.cause instanceof _BaseError ? args.cause.docsPath || args.docsPath : args.docsPath;
    this._tag = _tag;
    this.name = _tag;
    if (details) {
      this.details = /** @type {string}*/
      details;
    }
    this.docsPath = docsPath;
    this.metaMessages = args.metaMessages;
    this.shortMessage = shortMessage;
    this.version = getVersion();
    this.code = code;
    if (args.cause) {
      this.cause = /**@type {any} */
      args.cause;
    }
    this.message = [
      shortMessage || "An error occurred.",
      "",
      ...args.metaMessages ? [...args.metaMessages, ""] : [],
      ...docsPath ? [`Docs: ${args.docsBaseUrl ?? "https://tevm.sh"}${docsPath}${args.docsSlug ? `#${args.docsSlug}` : ""}`] : [],
      ...details ? [`Details: ${details}`] : [],
      `Version: ${this.version}`
    ].join("\n");
  }
  /**
   * Walks through the error chain.
   * @param {Function} [fn] - A function to execute on each error in the chain.
   * @returns {Error|unknown} The first error that matches the function, or the original error.
   */
  walk(fn) {
    return walk(this, fn);
  }
};
function walk(err, fn) {
  if (fn?.(err)) return err;
  if (err && typeof err === "object" && "cause" in err) return walk(err.cause, fn);
  return fn ? null : err;
}

// src/ethereum/InternalErrorError.js
var InternalError = class _InternalError extends BaseError {
  /**
   * The error code for InternalError.
   * @type {number}
   */
  static code = -32603;
  /**
   * Constructs an InternalError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InternalErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InternalError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InternalError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/internalerror/"
      },
      tag,
      _InternalError.code
    );
    this.meta = args.meta;
  }
};

// src/client/MisconfiguredClient.js
var MisconfiguredClientError = class extends InternalError {
  /**
   * Constructs a MisconfiguredClientError.
   *
   * @param {string} [message='Misconfigured client error occurred.'] - Human-readable error message.
   * @param {MisconfiguredClientErrorParameters} [args={}] - Additional parameters for the error.
   */
  constructor(message = "Misconfigured client error occurred.", args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/misconfiguredclienterror/"
      },
      "MisconfiguredClientError"
    );
    this.name = "MisconfiguredClientError";
    this._tag = "MisconfiguredClientError";
  }
};

// src/ethereum/ExecutionErrorError.js
var ExecutionError = class _ExecutionError extends BaseError {
  /**
   * The error code for ExecutionError.
   * @type {number}
   */
  static code = -32015;
  /**
   * Constructs an ExecutionError.
   *
   * @param {string} message - Human-readable error message.
   * @param {ExecutionErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='ExecutionError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "ExecutionError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/executionerror/"
      },
      tag,
      _ExecutionError.code
    );
  }
};

// src/common/CommonMismatchError.js
var CommonMismatchError = class extends ExecutionError {
  /**
   * Constructs a CommonMismatchError.
   *
   * @param {string} [message='Common mismatch error occurred.'] - Human-readable error message.
   * @param {CommonMismatchErrorParameters} [args={}] - Additional parameters for the error.
   */
  constructor(message = "Common mismatch error occurred.", args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/commonmismatcherror/"
      },
      "CommonMismatchError"
    );
  }
};

// src/common/EipNotEnabledError.js
var EipNotEnabledError = class extends ExecutionError {
  /**
   * Constructs an EipNotEnabledError.
   *
   * @param {string} [message='EIP not enabled error occurred.'] - Human-readable error message.
   * @param {EipNotEnabledErrorParameters} [args={}] - Additional parameters for the error.
   */
  constructor(message = "EIP not enabled error occurred.", args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/eipnotenablederror/"
      },
      "EipNotEnabledError"
    );
    this.name = "EipNotEnabledError";
    this._tag = "EipNotEnabledError";
  }
};

// src/data/InvalidByteSizeError.js
var InvalidBytesSizeError = class extends InternalError {
  /**
   * The actual size of the bytes.
   * @type {number}
   */
  size;
  /**
   * The expected size of the bytes.
   * @type {number}
   */
  expectedSize;
  /**
   * Constructs an InvalidBytesSizeError.
   *
   * @param {number} size - The actual size of the bytes.
   * @param {number} expectedSize - The expected size of the bytes.
   * @param {string} [message] - Human-readable error message.
   * @param {InvalidBytesSizeErrorParameters} [args] - Additional parameters for the error.
   */
  constructor(size, expectedSize, message, args = {}) {
    super(
      message || `Invalid bytes size. Expected ${expectedSize}, received ${size}`,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidbytessizeerror/"
      },
      "InvalidBytesSizeError"
    );
    this.size = size;
    this.expectedSize = expectedSize;
    this.name = "InvalidBytesSizeError";
    this._tag = "InvalidBytesSizeError";
  }
};

// src/defensive/DefensiveNullCheckError.js
var DefensiveNullCheckError = class extends InternalError {
  /**
   * Constructs a DefensiveNullCheckError.
   *
   * @param {string} [message] - Human-readable error message.
   * @param {DefensiveNullCheckErrorParameters} [args] - Additional parameters for the error.
   */
  constructor(message, args = {}) {
    super(
      message || "Defensive null check error occurred.",
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/defensivenullcheckerror/"
      },
      "DefensiveNullCheckError"
    );
    this.name = "DefensiveNullCheckError";
    this._tag = "DefensiveNullCheckError";
  }
};

// src/defensive/UnreachableCodeError.js
var UnreachableCodeError = class extends InternalError {
  /**
   * The value that should be unreachable.
   * @type {*}
   */
  value;
  /**
   * Constructs an UnreachableCodeError.
   *
   * @param {*} value - The value that should be unreachable.
   * @param {string} [message] - Human-readable error message.
   * @param {UnreachableCodeErrorParameters} [args] - Additional parameters for the error.
   */
  constructor(value, message, args = {}) {
    super(
      message || `Unreachable code executed with value: ${JSON.stringify(value)}`,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/unreachablecodeerror/",
        details: `Unreachable value: ${JSON.stringify(value)}`
      },
      "UnreachableCodeError"
    );
    this.value = value;
    this.name = "UnreachableCodeError";
    this._tag = "UnreachableCodeError";
  }
};

// src/ethereum/AccountLockedError.js
var AccountLockedError = class _AccountLockedError extends BaseError {
  /**
   * The error code for AccountLockedError.
   * @type {number}
   */
  static code = -32020;
  /**
   * Constructs an AccountLockedError.
   *
   * @param {string} message - Human-readable error message.
   * @param {AccountLockedErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='AccountLocked'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "AccountLocked") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/accountlockederror/"
      },
      tag,
      _AccountLockedError.code
    );
  }
};

// src/ethereum/ResourceNotFoundError.js
var ResourceNotFoundError = class _ResourceNotFoundError extends BaseError {
  /**
   * The error code for ResourceNotFoundError.
   * @type {number}
   */
  static code = -32001;
  /**
   * Constructs a ResourceNotFoundError.
   *
   * @param {string} message - Human-readable error message.
   * @param {ResourceNotFoundErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='ResourceNotFound'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "ResourceNotFound") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/resourcenotfounderror/"
      },
      tag,
      _ResourceNotFoundError.code
    );
  }
};

// src/ethereum/AccountNotFoundError.js
var AccountNotFoundError = class extends ResourceNotFoundError {
  /**
   * Constructs an AccountNotFoundError.
   *
   * @param {string} message - Human-readable error message.
   * @param {AccountNotFoundErrorParameters} [args={}] - Additional parameters for the ResourceNotFoundError.
   * @param {string} [tag='AccountNotFound'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "AccountNotFound") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/accountnotfounderror/"
      },
      tag
    );
  }
};

// src/ethereum/BlobGasLimitExceededError.js
var BlobGasLimitExceededError = class _BlobGasLimitExceededError extends BaseError {
  /**
   * The error code for BlobGasLimitExceededError.
   * @type {number}
   */
  static code = -32003;
  /**
   * Constructs a BlobGasLimitExceededError.
   *
   * @param {string} [message='Blob gas limit exceeded'] - Human-readable error message.
   * @param {BlobGasLimitExceededErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='BlobGasLimitExceededError'] - The tag for the error.
   */
  constructor(message = "Blob gas limit exceeded", args = {}, tag = "BlobGasLimitExceededError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/blobgaslimitexceedederror/"
      },
      tag,
      _BlobGasLimitExceededError.code
    );
  }
};

// src/ethereum/BlockGasLimitExceededError.js
var BlockGasLimitExceededError = class _BlockGasLimitExceededError extends BaseError {
  /**
   * The error code for BlockGasLimitExceededError.
   * @type {number}
   */
  static code = -32006;
  /**
   * Constructs a BlockGasLimitExceededError.
   *
   * @param {string} message - Human-readable error message.
   * @param {BlockGasLimitExceededErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='BlockGasLimitExceeded'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "BlockGasLimitExceeded") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/blockgaslimitexceedederror/"
      },
      tag,
      _BlockGasLimitExceededError.code
    );
    this.meta = args.meta;
  }
};

// src/ethereum/ChainIdMismatchError.js
var ChainIdMismatchError = class _ChainIdMismatchError extends BaseError {
  /**
   * The error code for ChainIdMismatchError.
   * @type {number}
   */
  static code = -32e3;
  /**
   * Constructs a ChainIdMismatchError.
   *
   * @param {string} message - Human-readable error message.
   * @param {ChainIdMismatchErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='ChainIdMismatch'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "ChainIdMismatch") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/chainidmismatcherror/"
      },
      tag,
      _ChainIdMismatchError.code
    );
  }
};

// src/ethereum/ContractExecutionFailedError.js
var ContractExecutionFailedError = class _ContractExecutionFailedError extends BaseError {
  /**
   * The error code for ContractExecutionFailedError.
   * @type {number}
   */
  static code = -32015;
  /**
   * Constructs a ContractExecutionFailedError.
   *
   * @param {string} message - Human-readable error message.
   * @param {ContractExecutionFailedErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='ContractExecutionFailed'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "ContractExecutionFailed") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/contractexecutionfailederror/"
      },
      tag,
      _ContractExecutionFailedError.code
    );
  }
};

// src/ethereum/ethereumjs/AuthCallUnsetError.js
var AuthCallUnsetError = class extends ExecutionError {
  // Note: AUTHCALL_UNSET was removed from ethereumjs
  /** @type {string} */
  // static EVMErrorMessage = EVMError.errorMessages.AUTHCALL_UNSET
  /**
   * Constructs an AuthCallUnsetError.
   * Represents an EIP-3074 specific error that occurs when attempting to AUTHCALL without AUTH set.
   *
   * AuthCallUnset errors can occur due to:
   * - Attempting to execute an AUTHCALL without setting the necessary authorization.
   *
   * To debug an AuthCallUnset error:
   * 1. **Review Authorization Logic**: Ensure that the necessary authorization is set before executing an AUTHCALL.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the AUTHCALL is attempted without AUTH set.
   *
   * @param {string} [message='AuthCallUnset error occurred.'] - Human-readable error message.
   * @param {AuthCallUnsetErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='AuthCallUnsetError'] - The tag for the error.
   */
  constructor(message = "AuthCallUnset error occurred.", args = {}, tag = "AuthCallUnsetError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/authcallunseterror/"
      },
      tag
    );
    this.message = message;
    this.meta = args.meta;
    this._tag = "AuthCallUnsetError";
  }
};
var BLS12381FpNotInFieldError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.BLS_12_381_FP_NOT_IN_FIELD;
  /**
   * Constructs a BLS12381FpNotInFieldError.
   * Represents an EIP-2537 specific error that occurs when an fp point is not in the field during BLS12-381 operations.
   *
   * Fp point not in field errors can occur due to:
   * - Providing an fp point that does not lie within the expected field for BLS12-381 operations.
   *
   *
   * @param {string} [message='BLS12-381 fp point not in field error occurred.'] - Human-readable error message.
   * @param {BLS12381FpNotInFieldErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='BLS12381FpNotInFieldError'] - The tag for the error.
   */
  constructor(message = "BLS12-381 fp point not in field error occurred.", args = {}, tag = "BLS12381FpNotInFieldError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/bls12381fpnotinfielderror/"
      },
      tag
    );
    this.message = message;
    this.meta = args.meta;
  }
};
var BLS12381InputEmptyError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.BLS_12_381_INPUT_EMPTY;
  /**
   * Constructs a BLS12381InputEmptyError.
   * Represents an EIP-2537 specific error that occurs when an input is empty during BLS12-381 operations.
   *
   * Input empty errors can occur due to:
   * - Providing empty input data for BLS12-381 operations.
   *
   *
   * @param {string} [message='BLS12-381 input empty error occurred.'] - Human-readable error message.
   * @param {BLS12381InputEmptyErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='BLS12381InputEmptyError'] - The tag for the error.
   */
  constructor(message = "BLS12-381 input empty error occurred.", args = {}, tag = "BLS12381InputEmptyError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/bls12381inputemptyerror/"
      },
      tag
    );
  }
};
var BLS12381InvalidInputLengthError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.BLS_12_381_INVALID_INPUT_LENGTH;
  /**
   * Constructs a BLS12381InvalidInputLengthError.
   * Represents an EIP-2537 specific error that occurs when an invalid input length is encountered during BLS12-381 operations.
   *
   * Invalid input length errors can occur due to:
   * - Providing input data of incorrect length for BLS12-381 operations.
   *
   *
   * @param {string} [message='BLS12-381 invalid input length error occurred.'] - Human-readable error message.
   * @param {BLS12381InvalidInputLengthErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='BLS12381InvalidInputLengthError'] - The tag for the error.
   */
  constructor(message = "BLS12-381 invalid input length error occurred.", args = {}, tag = "BLS12381InvalidInputLengthError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/bls12381invalidinputlengtherror/"
      },
      tag
    );
  }
};
var BLS12381PointNotOnCurveError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.BLS_12_381_POINT_NOT_ON_CURVE;
  /**
   * Constructs a BLS12381PointNotOnCurveError.
   * Represents an EIP-2537 specific error that occurs when a point is not on the curve during BLS12-381 operations.
   *
   * Point not on curve errors can occur due to:
   * - Providing a point that does not lie on the expected curve for BLS12-381 operations.
   *
   * @param {string} [message='BLS12-381 point not on curve error occurred.'] - Human-readable error message.
   * @param {BLS12381PointNotOnCurveErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='BLS12381PointNotOnCurveError'] - The tag for the error.
   */
  constructor(message = "BLS12-381 point not on curve error occurred.", args = {}, tag = "BLS12381PointNotOnCurveError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/bls12381pointnotoncurveerror/"
      },
      tag
    );
  }
};

// src/ethereum/GasLimitExceededError.js
var GasLimitExceededError = class _GasLimitExceededError extends BaseError {
  /**
   * The error code for GasLimitExceededError.
   * @type {number}
   */
  static code = -32003;
  /**
   * Constructs a GasLimitExceededError.
   *
   * @param {string} message - Human-readable error message.
   * @param {GasLimitExceededErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='GasLimitExceeded'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "GasLimitExceeded") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/gaslimitexceedederror/"
      },
      tag,
      _GasLimitExceededError.code
    );
  }
};

// src/ethereum/ethereumjs/CodeSizeExceedsMaximumError.js
var CodeSizeExceedsMaximumError = class extends GasLimitExceededError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.CODESIZE_EXCEEDS_MAXIMUM;
  /**
   * Constructs a CodeSizeExceedsMaximumError.
   * Represents an calldata/creation error that occurs when the code size exceeds the maximum limit.
   * This error is typically encountered when the contract size to be deployed exceeds the maximum allowed size.
   *
   * Code size exceeds maximum errors can occur due to:
   * - Deployment of contracts with large bytecode.
   * - Contracts with a significant amount of embedded data or logic.
   * - Incorrect settings for contract size limits in TEVM configuration.
   *
   * To debug a code size exceeds maximum error:
   * 1. **Review Contract Size**: Ensure that the contract bytecode size is within the allowed limits. Consider refactoring the contract to reduce its size.
   * 2. **Optimize Contract Code**: Break down large contracts into smaller, modular contracts and use libraries or inheritance to share code.
   * 3. **Configure TEVM Memory Client**: When creating a TEVM MemoryClient instance, set `allowUnlimitedContractSize` to `true` if necessary. Note that even with this setting, you may still encounter block limits.
   *    ```typescript
   *    import { createMemoryClient } from 'tevm'
   *
   *    const client = createMemoryClient({ allowUnlimitedContractSize: true })
   *    ```
   * 4. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract deployment process and inspect the bytecode size.
   * 5. **Use Other Tools**: Use other tools to analyze and optimize contract bytecode.
   *
   *
   * @param {string} [message='Code size exceeds maximum error occurred.'] - Human-readable error message.
   * @param {CodeSizeExceedsMaximumErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='CodeSizeExceedsMaximumError'] - The tag for the error.
   */
  constructor(message = "Code size exceeds maximum error occurred.", args = {}, tag = "CodeSizeExceedsMaximumError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/codesizeexceedsmaximumerror/"
      },
      tag
    );
  }
};
var CodeStoreOutOfGasError = class extends GasLimitExceededError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.CODESTORE_OUT_OF_GAS;
  /**
   * Constructs a CodeStoreOutOfGasError.
   * Represents an error that occurs when a transaction runs out of gas during code storage.
   * This error is typically encountered when the gas provided for storing code is insufficient to complete its execution.
   * EVM transaction execution metadata level error
   *
   * Code store out of gas errors can occur due to:
   * - Insufficient gas provided for storing large contracts.
   * - Incorrect estimation of gas required for storing code.
   * - Contracts with high gas consumption during the deployment phase.
   * - Non-deterministic gas usage during code storage.
   * - If TEVM submitted the transaction using `createTransaction: true` and the account being used runs out of gas.
   *
   * @param {string} [message='Code store out of gas error occurred.'] - Human-readable error message.
   * @param {CodeStoreOutOfGasErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='CodeStoreOutOfGasError'] - The tag for the error.
   */
  constructor(message = "Code store out of gas error occurred.", args = {}, tag = "CodeStoreOutOfGasError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/codestoreoutofgaserror/"
      },
      tag
    );
  }
};
var CreateCollisionError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.CREATE_COLLISION;
  /**
   * Constructs a CreateCollisionError.
   * Represents an execution error that occurs when a contract creation results in a collision.
   *
   * Create collision errors can occur due to:
   * - Attempting to deploy a contract to an address that is already in use.
   *
   * To debug a create collision error:
   * 1. **Review Deployment Logic**: Ensure that the contract address is not already in use.
   * 2. **Nonces** Check that the nonce of the account used had been incremented. Remember nonces do not update until a block is mined.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the deployment process and identify the collision.
   *
   * @param {string} [message='Create collision error occurred.'] - Human-readable error message.
   * @param {CreateCollisionErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='CreateCollisionError'] - The tag for the error.
   */
  constructor(message = "Create collision error occurred.", args = {}, tag = "CreateCollisionError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/createcollisionerror/"
      },
      tag
    );
  }
};

// src/ethereum/RevertError.js
var RevertError = class _RevertError extends BaseError {
  /**
   * The error code for RevertError.
   * @type {number}
   */
  static code = 3;
  /**
   * Constructs a RevertError.
   *
   * @param {string} message - Human-readable error message.
   * @param {RevertErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='Revert'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "Revert") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/reverterror/"
      },
      tag,
      _RevertError.code
    );
    if (args.raw) {
      this.raw = args.raw;
    }
  }
};

// src/ethereum/ethereumjs/EvmRevertError.js
var EvmRevertError = class extends RevertError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.REVERT;
  /**
   * Constructs a EvmRevertError.
   * Represents an execution error that occurs when a transaction is reverted during EVM execution.
   * This error is typically encountered when a smart contract execution is reverted due to unmet conditions or failed assertions.
   *
   * EvmRevert errors can occur due to:
   * - Failed assertions in the smart contract code.
   * - Conditions in the code that trigger a revert.
   * - Insufficient gas to complete the transaction.
   * - Contract logic that intentionally reverts under certain conditions.
   *
   * To debug a revert error:
   * 1. **Review Revert Conditions**: Ensure that the conditions in the contract code that trigger reverts are properly handled and expected.
   * 2. **Check Assertions**: Verify that all assertions in the code are valid and necessary.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the revert occurs.
   * 4. **Inspect Contract Logic**: Manually inspect the contract code to understand why the revert is being triggered and ensure it is intentional.
   *
   * @param {string} [message='Revert error occurred.'] - Human-readable error message.
   * @param {EvmRevertErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='EvmRevertError'] - The tag for the error.
   */
  constructor(message = "Revert error occurred.", args = {}, tag = "EvmRevertError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/reverterror/"
      },
      tag
    );
  }
};
var InitcodeSizeViolationError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INITCODE_SIZE_VIOLATION;
  /**
   * Constructs an InitcodeSizeViolationError.
   * Represents a calldata/creation error that occurs when initcode exceeds the maximum allowable size during EVM execution.
   *
   * Initcode size violation errors can occur due to:
   * - Bugs in the smart contract code causing the initcode to exceed the maximum size.
   * - Issues during the deployment process resulting in oversized initcode.
   *
   * To debug an initcode size violation error:
   * 1. **Review Deployment Process**: Ensure that the initcode being deployed is within the allowable size limits.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract deployment and identify where the initcode size violation occurs.
   *
   * @param {string} [message='Initcode size violation error occurred.'] - Human-readable error message.
   * @param {InitcodeSizeViolationErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InitcodeSizeViolationError'] - The tag for the error.
   */
  constructor(message = "Initcode size violation error occurred.", args = {}, tag = "InitcodeSizeViolationError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/initcodesizeviolationerror/"
      },
      tag
    );
  }
};
var InsufficientBalanceError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INSUFFICIENT_BALANCE;
  /**
   * Constructs an InsufficientBalanceError.
   * Represents an error that occurs when an account has insufficient balance to perform a transaction.
   * EVM transaction execution metadata level error
   *
   * Insufficient balance errors can occur due to:
   * - Attempting to transfer or spend more funds than available in the account.
   *
   * To debug an insufficient balance error:
   * 1. **Review Account Balance**: Ensure that the account has sufficient funds to cover the transaction.
   * 2. **Check Transaction Details**: Verify the transaction amount and ensure it does not exceed the account balance.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the transaction execution and identify where the insufficient balance occurs.
   *
   * @param {string} [message='Insufficient balance error occurred.'] - Human-readable error message.
   * @param {InsufficientBalanceErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InsufficientBalanceError'] - The tag for the error.
   */
  constructor(message = "Insufficient balance error occurred.", args = {}, tag = "InsufficientBalanceError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/insufficientbalanceerror/"
      },
      tag
    );
  }
};
var InternalEvmError = class extends BaseError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INTERNAL_ERROR;
  /**
   * Constructs an InternalEvmError.
   * Represents an internal error within the EVM. This error should not typically happen
   * This error is typically encountered when there is an unexpected issue within the EVM execution or client.
   *
   * Internal errors can occur due to:
   * - Bugs in the EVM implementation.
   * - Bugs in Tevm
   *
   * If you encounter this error please open an issue on the Tevm GitHub repository.
   *
   * @param {string} [message='Internal error occurred.'] - Human-readable error message.
   * @param {InternalEvmErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InternalEvmError'] - The tag for the error.
   */
  constructor(message = "Internal error occurred.", args = {}, tag = "InternalEvmError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/internalerror/"
      },
      tag
    );
  }
};

// src/ethereum/ethereumjs/InvalidBeginSubError.js
var InvalidBeginSubError = class extends ExecutionError {
  // Note: INVALID_BEGINSUB was removed from ethereumjs
  /** @type {string} */
  // static EVMErrorMessage = EVMError.errorMessages.INVALID_BEGINSUB
  /**
   * Constructs an InvalidBeginSubError.
   * Represents an invalid bytecode/contract error that occurs when an invalid BEGINSUB operation is executed within the EVM.
   *
   * Invalid BEGINSUB errors can occur due to:
   * - Incorrect use of the BEGINSUB opcode.
   * - Bugs in the smart contract code causing invalid subroutine execution.
   *
   * To debug an invalid BEGINSUB error:
   * 1. **Review Subroutine Logic**: Ensure that the BEGINSUB opcode is used correctly within subroutine definitions.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the invalid BEGINSUB occurs.
   *
   * @param {string} [message='Invalid BEGINSUB error occurred.'] - Human-readable error message.
   * @param {InvalidBeginSubErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidBeginSubError'] - The tag for the error.
   */
  constructor(message = "Invalid BEGINSUB error occurred.", args = {}, tag = "InvalidBeginSubError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidbeginsuberror/"
      },
      tag
    );
  }
};
var InvalidBytecodeResultError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_BYTECODE_RESULT;
  /**
   * Constructs an InvalidBytecodeResultError.
   * Represents a calldata/creation error that occurs when invalid bytecode is deployed during EVM execution.
   *
   * Invalid bytecode result errors can occur due to:
   * - Bugs in the smart contract code causing invalid bytecode to be generated.
   * - Issues during the deployment process resulting in invalid bytecode.
   *
   * To debug an invalid bytecode result error:
   * 1. **Review Deployment Process**: Ensure that the bytecode being deployed is valid and correctly generated.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract deployment and identify where the invalid bytecode is generated or deployed.
   *
   * @param {string} [message='Invalid bytecode result error occurred.'] - Human-readable error message.
   * @param {InvalidBytecodeResultErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidBytecodeResultError'] - The tag for the error.
   */
  constructor(message = "Invalid bytecode result error occurred.", args = {}, tag = "InvalidBytecodeResultError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidbytecoderesulterror/"
      },
      tag
    );
  }
};
var InvalidCommitmentError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_COMMITMENT;
  /**
   * Constructs an InvalidCommitmentError.
   * Represents an EIP-4844 specific error that occurs when a KZG commitment does not match the versioned hash.
   *
   * Invalid commitment errors can occur due to:
   * - Providing a KZG commitment that does not match the expected versioned hash.
   *
   * @example
   * ```typescript
   * import { InvalidCommitmentError } from '@tevm/errors'
   * try {
   *   // Some operation that can throw an InvalidCommitmentError
   * } catch (error) {
   *   if (error instanceof InvalidCommitmentError) {
   *     console.error(error.message);
   *     // Handle the invalid commitment error
   *   }
   * }
   * ```
   *
   * @param {string} [message='Invalid commitment error occurred.'] - Human-readable error message.
   * @param {InvalidCommitmentErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidCommitmentError'] - The tag for the error.
   */
  constructor(message = "Invalid commitment error occurred.", args = {}, tag = "InvalidCommitmentError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidcommitmenterror/"
      },
      tag
    );
  }
};
var InvalidEofFormatError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_EOF_FORMAT;
  /**
   * Constructs an InvalidEofFormatError.
   * Represents an error that occurs when an invalid EOF format is encountered during EVM execution.
   *
   * This error is specific to EOF
   *
   * Invalid EOF format errors can occur due to:
   * - Bugs in the smart contract code causing invalid EOF format.
   * - Issues during the deployment process resulting in invalid EOF format.
   *
   * To debug an invalid EOF format error:
   * 1. **Review Deployment Process**: Ensure that the EOF format being used is valid and correctly generated.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract deployment and identify where the invalid EOF format is generated or deployed.
   *
   * @param {string} [message='Invalid EOF format error occurred.'] - Human-readable error message.
   * @param {InvalidEofFormatErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidEofFormatError'] - The tag for the error.
   */
  constructor(message = "Invalid EOF format error occurred.", args = {}, tag = "InvalidEofFormatError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalideofformatterror/"
      },
      tag
    );
  }
};
var InvalidInputLengthError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_INPUT_LENGTH;
  /**
   * Constructs an InvalidInputLengthError.
   * Represents a calldata/creation error that occurs when an invalid input length is encountered during EVM execution.
   *
   * Invalid input length errors can occur due to:
   * - Providing input data of incorrect length.
   *
   * To debug an invalid input length error:
   * 1. **Review Input Data**: Ensure that the input data provided matches the expected length.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the invalid input length is encountered.
   *
   * @param {string} [message='Invalid input length error occurred.'] - Human-readable error message.
   * @param {InvalidInputLengthErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidInputLengthError'] - The tag for the error.
   */
  constructor(message = "Invalid input length error occurred.", args = {}, tag = "InvalidInputLengthError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidinputlengtherror/"
      },
      tag
    );
  }
};
var InvalidJumpError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_JUMP;
  /**
   * Constructs an InvalidJumpError.
   * Represents an invalid bytecode/contract error that occurs when an invalid JUMP operation is executed within the EVM.
   * This error is typically encountered when the jump destination in the bytecode is invalid or does not exist.
   *
   * Invalid JUMP errors can occur due to:
   * - Incorrect jump destinations in the bytecode.
   * - Bugs in the smart contract code causing jumps to non-existent locations.
   * - Conditional logic errors leading to unexpected jump destinations.
   *
   * To debug an invalid JUMP error:
   * 1. **Double Check Bytecode**: Ensure that the bytecode provided is valid.
   * 2. **Verify Common Configuration**: Ensure you are using a `common` with the correct hardfork and EIPs that support the EVM version you are targeting.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the bytecode execution and identify where the invalid JUMP occurs.
   *
   * @param {string} [message='Invalid JUMP error occurred.'] - Human-readable error message.
   * @param {InvalidJumpErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag] - Optionally override the name/tag for the error.
   * @param {string} [tag='InvalidJumpError'] - The tag for the error.}
   */
  constructor(message = "Invalid JUMP error occurred.", args = {}, tag = "InvalidJumpError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidjumperror/"
      },
      tag
    );
  }
};

// src/ethereum/ethereumjs/InvalidJumpSubError.js
var InvalidJumpSubError = class extends ExecutionError {
  // Note: INVALID_JUMPSUB was removed from ethereumjs
  /** @type {string} */
  // static EVMErrorMessage = EVMError.errorMessages.INVALID_JUMPSUB
  /**
   * Constructs an InvalidJumpSubError.
   * Represents an invalid bytecode/contract error that occurs when an invalid JUMPSUB operation is executed within the EVM.
   *
   * Invalid JUMPSUB errors can occur due to:
   * - Incorrect use of the JUMPSUB opcode.
   * - Bugs in the smart contract code causing invalid subroutine jumps.
   *
   * To debug an invalid JUMPSUB error:
   * 1. **Review Subroutine Logic**: Ensure that the JUMPSUB opcode is used correctly within subroutine definitions.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the invalid JUMPSUB occurs.
   *
   * @param {string} [message='Invalid JUMPSUB error occurred.'] - Human-readable error message.
   * @param {InvalidJumpSubErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidJumpSubError'] - The tag for the error.}
   *
   * @example
   * ```typescript
   * import { InvalidJumpSubError } from '@tevm/errors'
   * try {
   *   // Some operation that can throw an InvalidJumpSubError
   * } catch (error) {
   *   if (error instanceof InvalidJumpSubError) {
   *     console.error(error.message);
   *     // Handle the invalid JUMPSUB error
   *   }
   * }
   * ```
   */
  constructor(message = "Invalid JUMPSUB error occurred.", args = {}, tag = "InvalidJumpSubError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidjumpsuberror/"
      },
      tag
    );
  }
};
var InvalidKzgInputsError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_INPUTS;
  /**
   * Constructs an InvalidKzgInputsError.
   * Represents an EIP-4844 specific error that occurs when KZG inputs are invalid.
   *
   * Invalid inputs errors can occur due to:
   * - Providing invalid KZG inputs that do not meet the expected criteria.
   *
   *
   * @param {string} [message='Invalid inputs error occurred.'] - Human-readable error message.
   * @param {InvalidKzgInputsErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidKzgInputsError'] - The tag for the error.}
   */
  constructor(message = "Invalid inputs error occurred.", args = {}, tag = "InvalidKzgInputsError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidinputserror/"
      },
      tag
    );
  }
};
var InvalidOpcodeError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_OPCODE;
  /**
   * Constructs an InvalidOpcodeError.
   * Represents an invalid bytecode/contract error that occurs when an invalid opcode is encountered during EVM execution.
   * This error is typically encountered when the bytecode contains an opcode that is not recognized by the EVM.
   *
   * Invalid opcode errors can occur due to:
   * - Typographical errors in the bytecode.
   * - Using opcodes that are not supported by the selected EVM version or hardfork.
   *
   * To debug an invalid opcode error:
   * 1. **Review Bytecode**: Ensure that the bytecode provided is correct and does not contain any invalid opcodes.
   * 2. **Verify Common Configuration**: Ensure you are using a `common` with the correct hardfork and EIPs that support the opcodes used by your contract.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the bytecode execution and identify where the invalid opcode is encountered.
   * 4. **Inspect Contract Code**: Manually inspect the contract code to ensure it compiles correctly and does not include any invalid opcodes.
   *
   * @param {string} [message='Invalid opcode error occurred.'] - Human-readable error message.
   * @param {InvalidOpcodeErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidOpcodeError'] - The tag for the error.}
   */
  constructor(message = "Invalid opcode error occurred.", args = {}, tag = "InvalidOpcodeError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidopcodeerror/"
      },
      tag
    );
  }
};
var InvalidProofError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.INVALID_PROOF;
  /**
   * Constructs an InvalidProofError.
   * Represents an EIP-4844 specific error that occurs when a KZG proof is invalid.
   *
   * Invalid proof errors can occur due to:
   * - Providing a KZG proof that does not meet the expected criteria.
   *
   * @param {string} [message='Invalid proof error occurred.'] - Human-readable error message.
   * @param {InvalidProofErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidProofError'] - The tag for the error.
   */
  constructor(message = "Invalid proof error occurred.", args = {}, tag = "InvalidProofError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidprooferror/"
      },
      tag
    );
  }
};

// src/ethereum/ethereumjs/InvalidReturnSubError.js
var InvalidReturnSubError = class extends ExecutionError {
  // Note: INVALID_RETURNSUB was removed from ethereumjs
  /** @type {string} */
  // static EVMErrorMessage = EVMError.errorMessages.INVALID_RETURNSUB
  /**
   * Constructs an InvalidReturnSubError.
   * Represents an invalid bytecode/contract error that occurs when an invalid RETURNSUB operation is executed within the EVM.
   *
   * Invalid RETURNSUB errors can occur due to:
   * - Incorrect use of the RETURNSUB opcode.
   * - Bugs in the smart contract code causing invalid subroutine returns.
   *
   * To debug an invalid RETURNSUB error:
   * 1. **Review Subroutine Logic**: Ensure that the RETURNSUB opcode is used correctly within subroutine definitions.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the invalid RETURNSUB occurs.
   *
   * @param {string} [message='Invalid RETURNSUB error occurred.'] - Human-readable error message.
   * @param {InvalidReturnSubErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidReturnSubError'] - The tag for the error.
   */
  constructor(message = "Invalid RETURNSUB error occurred.", args = {}, tag = "InvalidReturnSubError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidreturnsuberror/"
      },
      tag
    );
  }
};
var OutOfGasError = class extends GasLimitExceededError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.OUT_OF_GAS;
  /**
   * Constructs an OutOfGasError.
   * Represents an execution error that occurs when a transaction runs out of gas during execution.
   * This error is typically encountered when the gas provided for a transaction is insufficient to complete its execution.
   *
   * Out of gas errors can occur due to:
   * - Insufficient gas provided for complex transactions or loops.
   * - Incorrect estimation of gas required for certain operations.
   * - Contracts with high gas consumption in specific functions.
   * - Non-deterministic gas usage in contracts.
   * - If TEVM submitted the transaction using `createTransaction: true` and the account being used runs out of gas.
   *
   * To debug an out of gas error:
   * 1. **Review Gas Estimates**: Ensure that the gas estimate for your transaction is accurate and sufficient. If you provided explicit gas-related parameters, double-check their values.
   * 2. **Optimize Contract Code**: Refactor your smart contract code to reduce gas consumption, especially in loops and complex operations. Remove potential non-deterministic behaviors.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the transaction and inspect gas usage.
   * 4. **Estimate Gas Multiple Times**: If using TEVM gas estimations, it might make sense to estimate gas many times and take the worst case to set `gasPrice`. Most nodes execute `eth_estimateGas` 10 times, while TEVM runs it only once.
   * 5. **Use Other Tools**: Use other tools with gas profiling such as [Foundry](https://book.getfoundry.sh/forge/gas).
   *
   * @param {string} [message='Out of gas error occurred.'] - Human-readable error message.
   * @param {OutOfGasErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='OutOfGasError'] - The tag for the error.
   */
  constructor(message = "Out of gas error occurred.", args = {}, tag = "OutOfGasError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/outofgaserror/"
      },
      tag
    );
  }
};
var OutOfRangeError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.OUT_OF_RANGE;
  /**
   * Constructs an OutOfRangeError.
   * Represents an invalid bytecode/contract error that occurs when a value is out of the allowable range during EVM execution.
   * This error is typically encountered when an operation results in a value that exceeds the allowed limits.
   *
   * Value out of range errors can occur due to:
   * - Arithmetic operations that result in overflow or underflow.
   * - Incorrect handling of large numbers in the smart contract code.
   * - Bugs in the smart contract code causing values to exceed their expected range.
   *
   * To debug a value out of range error:
   * 1. **Review Arithmetic Operations**: Ensure that arithmetic operations in the contract are correctly handling large numbers and preventing overflow/underflow.
   * 2. **Check Value Assignments**: Verify that values assigned to variables are within the allowable range and properly validated.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the value goes out of range.
   * 4. **Inspect Contract Logic**: Manually inspect the contract code to ensure that all value assignments and operations are within the expected limits.
   *
   * @param {string} [message='Value out of range error occurred.'] - Human-readable error message.
   * @param {OutOfRangeErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='OutOfRangeError'] - The tag for the error.
   */
  constructor(message = "Value out of range error occurred.", args = {}, tag = "OutOfRangeError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/outofrangeerror/"
      },
      tag
    );
  }
};
var RefundExhaustedError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.REFUND_EXHAUSTED;
  /**
   * Constructs a RefundExhaustedError.
   * Represents an invalid bytecode error that occurs when the gas refund limit is exhausted.
   * EVM transaction execution metadata level error
   *
   * Refund exhausted errors can occur due to:
   * - The transaction exceeding the gas refund limit.
   *
   * To debug a refund exhausted error:
   * 1. **Review Gas Usage**: Ensure that the gas usage in the contract is within the refund limits.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify excessive gas usage.
   *
   *
   * @param {string} [message='Refund exhausted error occurred.'] - Human-readable error message.
   * @param {RefundExhaustedErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='RefundExhaustedError'] - The tag for the error.}
   */
  constructor(message = "Refund exhausted error occurred.", args = {}, tag = "RefundExhaustedError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/refundexhaustederror/"
      },
      tag
    );
  }
};
var StackOverflowError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.STACK_OVERFLOW;
  /**
   * Constructs a StackOverflowError.
   * Represents an invalid bytecode error that occurs when there is a stack overflow during execution.
   * This error is typically encountered when an operation causes the stack to exceed its limit.
   *
   * Stack overflow errors can occur due to:
   * - Excessive recursion leading to too many function calls.
   * - Bugs in smart contract logic that cause infinite loops or excessive stack usage.
   * - Incorrect management of stack operations (e.g., pushing too many items onto the stack).
   *
   * To debug a stack overflow error:
   * 1. **Review Contract Logic**: Ensure that your smart contract logic correctly handles recursion and stack operations.
   * 2. **Optimize Stack Usage**: Refactor your code to reduce stack usage, such as minimizing the depth of recursive calls.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the transaction and inspect stack changes.
   * 4. **Use Other Tools**: Use other tools with tracing such as [Foundry](https://book.getfoundry.sh/forge/traces).
   * - **Ethereumjs Source**: Refer to the [source file](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/evm/src/stack.ts) where this error can occur.
   *
   * @param {string} [message='Stack overflow error occurred.'] - Human-readable error message.
   * @param {StackOverflowErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='StackOverflowError'] - The tag for the error.
   */
  constructor(message = "Stack overflow error occurred.", args = {}, tag = "StackOverflowError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/stackoverflowerror/"
      },
      tag
    );
  }
};
var StackUnderflowError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.STACK_UNDERFLOW;
  /**
   * Constructs a StackUnderflowError.
   * This error is typically encountered when an operation requires more stack items than are present.
   *
   * Stack underflow errors can occur due to:
   * - Incorrect management of stack operations (e.g., popping more items than available).
   * - Bugs in smart contract logic leading to unexpected stack behavior.
   * - Issues with function calls that manipulate the stack incorrectly.
   *
   * To debug a stack underflow error:
   * 1. **Review Contract Logic**: Ensure that your smart contract logic correctly handles stack operations, especially in loops and conditional branches.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the transaction and inspect stack changes.
   * 3. **Use Other Tools**: Use other tools with tracing such as [Foundry](https://book.getfoundry.sh/forge/traces).
   * - **Ethereumjs Source**: Refer to the [source file](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/evm/src/stack.ts) where this error can occur.
   *
   * @param {string} [message='Stack underflow error occurred.'] - Human-readable error message.
   * @param {StackUnderflowErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='StackUnderflowError'] - The tag for the error.
   */
  constructor(message = "Stack underflow error occurred.", args = {}, tag = "StackUnderflowError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/stackunderflowerror/"
      },
      tag
    );
  }
};
var StaticStateChangeError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.STATIC_STATE_CHANGE;
  /**
   * Constructs a StaticStateChangeError.
   * Represents an invalid bytecode/contract error that occurs when a state-changing operation is attempted in a static context.
   * This error is typically encountered when a contract attempts to modify the state during a static call.
   *
   * Static state change errors can occur due to:
   * - Attempting to modify the state in a static call.
   * - Executing state-changing operations in a read-only context.
   * - Bugs in the smart contract code leading to unintended state changes.
   *
   * To debug a static state change error:
   * 1. **Review Contract Logic**: Ensure that state-changing operations are not executed in static calls or read-only contexts.
   * 2. **Check Function Modifiers**: Verify that the function modifiers and visibility settings are correctly applied to prevent state changes in static contexts.
   * 3. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the state change is attempted in a static context.
   * 4. **Inspect Contract Code**: Manually inspect the contract code to ensure that state changes are correctly controlled and executed only in appropriate contexts.
   *
   * @param {string} [message='Static state change error occurred.'] - Human-readable error message.
   * @param {StaticStateChangeErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='StaticStateChangeError'] - The tag for the error.
   */
  constructor(message = "Static state change error occurred.", args = {}, tag = "StaticStateChangeError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/staticstatechangeerror/"
      },
      tag
    );
  }
};
var StopError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.STOP;
  /**
   * Constructs a StopError.
   * Represents a contract/bytecode error that occurs when a STOP operation is executed.
   *
   * Stop errors can occur due to:
   * - The contract execution reaching a STOP opcode.
   *
   * To debug a stop error:
   * 1. **Review Contract Logic**: Ensure that the STOP opcode is used intentionally and correctly.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify the STOP operation.
   *
   * @param {string} [message='Stop error occurred.'] - Human-readable error message.
   * @param {StopErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='StopError'] - The tag for the error.
   *
   */
  constructor(message = "Stop error occurred.", args = {}, tag = "StopError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/stoperror/"
      },
      tag
    );
  }
};
var ValueOverflowError = class extends ExecutionError {
  /** @type {string} */
  static EVMErrorMessage = evm.EVMError.errorMessages.VALUE_OVERFLOW;
  /**
   * Constructs a ValueOverflowError.
   * Represents an invalid bytecode/contract error that occurs when a value overflow happens during EVM execution.
   *
   * Value overflow errors can occur due to:
   * - Arithmetic operations that exceed the maximum value limit.
   *
   * To debug a value overflow error:
   * 1. **Review Arithmetic Operations**: Ensure that arithmetic operations are correctly handling large numbers and preventing overflow.
   * 2. **Use TEVM Tracing**: Utilize TEVM tracing to step through the contract execution and identify where the overflow occurs.
   *
   * @param {string} [message='Value overflow error occurred.'] - Human-readable error message.
   * @param {ValueOverflowErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='ValueOverflowError'] - The tag for the error.
   */
  constructor(message = "Value overflow error occurred.", args = {}, tag = "ValueOverflowError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/valueoverflowerror/"
      },
      tag
    );
  }
};

// src/ethereum/InsufficientFundsError.js
var InsufficientFundsError = class _InsufficientFundsError extends BaseError {
  /**
   * The error code for InsufficientFundsError.
   * @type {number}
   */
  static code = -32003;
  /**
   * Constructs an InsufficientFundsError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InsufficientFundsErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InsufficientFunds'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InsufficientFunds") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/insufficientfundserror/"
      },
      tag,
      _InsufficientFundsError.code
    );
  }
};

// src/ethereum/InsufficientPermissionsError.js
var InsufficientPermissionsError = class _InsufficientPermissionsError extends BaseError {
  /**
   * The error code for InsufficientPermissionsError.
   * @type {number}
   */
  static code = -32009;
  /**
   * Constructs an InsufficientPermissionsError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InsufficientPermissionsErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InsufficientPermissions'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InsufficientPermissions") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/insufficientpermissionserror/"
      },
      tag,
      _InsufficientPermissionsError.code
    );
  }
};

// src/ethereum/InvalidAddressError.js
var InvalidAddressError = class _InvalidAddressError extends BaseError {
  /**
   * The error code for InvalidAddressError.
   * @type {number}
   */
  static code = -32013;
  /**
   * Constructs an InvalidAddressError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidAddressErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidAddress'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidAddress") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidaddresserror/"
      },
      tag,
      _InvalidAddressError.code
    );
  }
};

// src/ethereum/InvalidGasPriceError.js
var InvalidGasPriceError = class _InvalidGasPriceError extends BaseError {
  /**
   * The error code for InvalidGasPriceError.
   * @type {number}
   */
  static code = -32012;
  /**
   * Constructs an InvalidGasPriceError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidGasPriceErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidGasPrice'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidGasPrice") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidgaspriceerror/"
      },
      tag,
      _InvalidGasPriceError.code
    );
  }
};

// src/ethereum/InvalidParamsError.js
var InvalidParamsError = class _InvalidParamsError extends BaseError {
  /**
   * The error code for InvalidParamsError.
   * @type {number}
   */
  static code = -32602;
  /**
   * Constructs an InvalidParamsError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidParamsErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidParams'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidParams") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidparamserror/"
      },
      tag,
      _InvalidParamsError.code
    );
  }
};

// src/ethereum/InvalidRequestError.js
var InvalidRequestError = class _InvalidRequestError extends BaseError {
  /**
   * The error code for InvalidRequestError.
   * @type {number}
   */
  static code = -32600;
  /**
   * Constructs an InvalidRequestError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidRequestErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidRequest'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidRequest") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidrequesterror/"
      },
      tag,
      _InvalidRequestError.code
    );
  }
};

// src/ethereum/InvalidSignatureError.js
var InvalidSignatureError = class _InvalidSignatureError extends BaseError {
  /**
   * The error code for InvalidSignatureError.
   * @type {number}
   */
  static code = -32e3;
  /**
   * Constructs an InvalidSignatureError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidSignatureErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidSignature'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidSignature") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidsignatureerror/"
      },
      tag,
      _InvalidSignatureError.code
    );
  }
};

// src/ethereum/InvalidTransactionError.js
var InvalidTransactionError = class _InvalidTransactionError extends BaseError {
  /**
   * The error code for InvalidTransactionError.
   * @type {number}
   */
  static code = -32003;
  /**
   * Constructs an InvalidTransactionError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidTransactionErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='InvalidTransaction'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidTransaction") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidtransactionerror/"
      },
      tag,
      _InvalidTransactionError.code
    );
  }
};

// src/ethereum/LimitExceededError.js
var LimitExceededError = class _LimitExceededError extends BaseError {
  /**
   * The error code for LimitExceededError.
   * @type {number}
   */
  static code = -32005;
  /**
   * Constructs a LimitExceededError.
   *
   * @param {string} message - Human-readable error message.
   * @param {LimitExceededErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='LimitExceeded'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "LimitExceeded") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/limitexceedederror/"
      },
      tag,
      _LimitExceededError.code
    );
  }
};

// src/ethereum/MethodNotFoundError.js
var MethodNotFoundError = class _MethodNotFoundError extends BaseError {
  /**
   * The error code for MethodNotFoundError.
   * @type {number}
   */
  static code = -32601;
  /**
   * Constructs a MethodNotFoundError.
   *
   * @param {string} message - Human-readable error message.
   * @param {MethodNotFoundErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='MethodNotFound'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "MethodNotFound") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/methodnotfounderror/"
      },
      tag,
      _MethodNotFoundError.code
    );
  }
};

// src/ethereum/MethodNotSupportedError.js
var MethodNotSupportedError = class _MethodNotSupportedError extends BaseError {
  /**
   * The error code for MethodNotSupportedError.
   * @type {number}
   */
  static code = -32004;
  /**
   * Constructs a MethodNotSupportedError.
   *
   * @param {string} message - Human-readable error message.
   * @param {MethodNotSupportedErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='MethodNotSupported'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "MethodNotSupported") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/methodnotsupportederror/"
      },
      tag,
      _MethodNotSupportedError.code
    );
  }
};

// src/ethereum/NonceAlreadyUsedError.js
var NonceAlreadyUsedError = class _NonceAlreadyUsedError extends BaseError {
  /**
   * The error code for NonceAlreadyUsedError.
   * @type {number}
   */
  static code = -32008;
  /**
   * Constructs a NonceAlreadyUsedError.
   *
   * @param {string} message - Human-readable error message.
   * @param {NonceAlreadyUsedErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='NonceAlreadyUsed'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "NonceAlreadyUsed") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/noncealreadyusederror/"
      },
      tag,
      _NonceAlreadyUsedError.code
    );
  }
};

// src/ethereum/NonceTooHighError.js
var NonceTooHighError = class _NonceTooHighError extends BaseError {
  /**
   * The error code for NonceTooHighError.
   * @type {number}
   */
  static code = -32003;
  /**
   * Constructs a NonceTooHighError.
   *
   * @param {string} message - Human-readable error message.
   * @param {NonceTooHighErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='NonceTooHigh'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "NonceTooHigh") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/noncetoohigherror/"
      },
      tag,
      _NonceTooHighError.code
    );
  }
};

// src/ethereum/NonceTooLowError.js
var NonceTooLowError = class _NonceTooLowError extends BaseError {
  /**
   * The error code for NonceTooLowError.
   * @type {number}
   */
  static code = -32003;
  /**
   * Constructs a NonceTooLowError.
   *
   * @param {string} message - Human-readable error message.
   * @param {NonceTooLowErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='NonceTooLow'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "NonceTooLow") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/noncetoolowerror/"
      },
      tag,
      _NonceTooLowError.code
    );
  }
};

// src/ethereum/ParseErrorError.js
var ParseError = class _ParseError extends BaseError {
  /**
   * The error code for ParseError.
   * @type {number}
   */
  static code = -32700;
  /**
   * Constructs a ParseError.
   *
   * @param {string} message - Human-readable error message.
   * @param {ParseErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='ParseError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "ParseError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/parseerror/"
      },
      tag,
      _ParseError.code
    );
  }
};

// src/ethereum/PendingTransactionTimeoutError.js
var PendingTransactionTimeoutError = class _PendingTransactionTimeoutError extends BaseError {
  /**
   * The error code for PendingTransactionTimeoutError.
   * @type {number}
   */
  static code = -32002;
  /**
   * Constructs a PendingTransactionTimeoutError.
   *
   * @param {string} message - Human-readable error message.
   * @param {PendingTransactionTimeoutErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='PendingTransactionTimeout'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "PendingTransactionTimeout") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/pendingtransactiontimeouterror/"
      },
      tag,
      _PendingTransactionTimeoutError.code
    );
  }
};

// src/ethereum/RateLimitExceededError.js
var RateLimitExceededError = class _RateLimitExceededError extends BaseError {
  /**
   * The error code for RateLimitExceededError.
   * @type {number}
   */
  static code = -32005;
  /**
   * Constructs a RateLimitExceededError.
   *
   * @param {string} message - Human-readable error message.
   * @param {RateLimitExceededErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='RateLimitExceeded'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "RateLimitExceeded") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/ratelimitexceedederror/"
      },
      tag,
      _RateLimitExceededError.code
    );
  }
};

// src/ethereum/ResourceUnavailableError.js
var ResourceUnavailableError = class _ResourceUnavailableError extends BaseError {
  /**
   * The error code for ResourceUnavailableError.
   * @type {number}
   */
  static code = -32002;
  /**
   * Constructs a ResourceUnavailableError.
   *
   * @param {string} message - Human-readable error message.
   * @param {ResourceUnavailableErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='ResourceUnavailable'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "ResourceUnavailable") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/resourceunavailableerror/"
      },
      tag,
      _ResourceUnavailableError.code
    );
  }
};

// src/ethereum/TransactionRejectedError.js
var TransactionRejectedError = class _TransactionRejectedError extends BaseError {
  /**
   * The error code for TransactionRejectedError.
   * @type {number}
   */
  static code = -32003;
  /**
   * Constructs a TransactionRejectedError.
   *
   * @param {string} message - Human-readable error message.
   * @param {TransactionRejectedErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='TransactionRejected'] - The tag for the error.}
   */
  constructor(message, args = {}, tag = "TransactionRejected") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/transactionrejectederror/"
      },
      tag,
      _TransactionRejectedError.code
    );
    this.meta = args.meta;
  }
};

// src/ethereum/TransactionTooLargeError.js
var TransactionTooLargeError = class _TransactionTooLargeError extends BaseError {
  /**
   * The error code for TransactionTooLargeError.
   * @type {number}
   */
  static code = -32011;
  /**
   * Constructs a TransactionTooLargeError.
   *
   * @param {string} message - Human-readable error message.
   * @param {TransactionTooLargeErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='TransactionTooLarge'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "TransactionTooLarge") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/transactiontoolargeerror/"
      },
      tag,
      _TransactionTooLargeError.code
    );
  }
};

// src/ethereum/TransactionUnderpricedError.js
var TransactionUnderpricedError = class _TransactionUnderpricedError extends BaseError {
  /**
   * The error code for TransactionUnderpricedError.
   * @type {number}
   */
  static code = -32014;
  /**
   * Constructs a TransactionUnderpricedError.
   *
   * @param {string} message - Human-readable error message.
   * @param {TransactionUnderpricedErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='TransactionUnderpriced'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "TransactionUnderpriced") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/transactionunderpricederror/"
      },
      tag,
      _TransactionUnderpricedError.code
    );
  }
};

// src/ethereum/UnknownBlockError.js
var UnknownBlockError = class _UnknownBlockError extends BaseError {
  /**
   * The error code for UnknownBlockError.
   * @type {number}
   */
  static code = -32001;
  /**
   * Constructs an UnknownBlockError.
   *
   * @param {string} message - Human-readable error message.
   * @param {UnknownBlockErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='UnknownBlock'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "UnknownBlock") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/unknownblockerror/"
      },
      tag,
      _UnknownBlockError.code
    );
  }
};

// src/ethereum/UnsupportedChainError.js
var UnsupportedChainError = class _UnsupportedChainError extends BaseError {
  /**
   * The error code for UnsupportedChainError.
   * @type {number}
   */
  static code = -32007;
  /**
   * Constructs an UnsupportedChainError.
   *
   * @param {string} message - Human-readable error message.
   * @param {UnsupportedChainErrorParameters} [args={}] - Additional parameters for the BaseError.
   * @param {string} [tag='UnsupportedChain'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "UnsupportedChain") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/unsupportedchainerror/"
      },
      tag,
      _UnsupportedChainError.code
    );
  }
};

// src/fork/ForkError.js
var ForkError = class extends BaseError {
  /**
   * Constructs a ForkError.
   *
   * @param {string} message - Human-readable error message.
   * @param {ForkErrorParameters} args - Additional parameters for the error.
   */
  constructor(message, args) {
    const cause = args.cause instanceof Error ? args.cause : new BaseError(args.cause.message, {}, "unknown", Number(args.cause.code));
    super(
      [message, cause.message].filter(Boolean).join("\n"),
      {
        ...args,
        cause,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/forkerror/"
      },
      "ForkError",
      "code" in args.cause ? Number(args.cause.code) : new ResourceNotFoundError("").code
    );
    this.name = "ForkError";
    this._tag = "ForkError";
  }
};

// src/fork/NoForkUriSetError.js
var NoForkTransportSetError = class extends BaseError {
  /**
   * Constructs a NoForkTransportSetError.
   *
   * @param {string} message - Human-readable error message.
   * @param {NoForkTransportSetErrorParameters} [args] - Additional parameters for the error.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/noforktransportseterror/"
      },
      "NoForkTransportSetError"
    );
    this.name = "NoForkTransportSetError";
    this._tag = "NoForkTransportSetError";
  }
};

// src/fork/NoForkUrlSetError.js
var NoForkUrlSetError = class extends BaseError {
  /**
   * Constructs a NoForkUrlSetError.
   *
   * @param {string} message - Human-readable error message.
   * @param {NoForkUrlSetErrorParameters} [args] - Additional parameters for the error.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/noforkurlseterror/"
      },
      "NoForkUrlSetError"
    );
    this.name = "NoForkUrlSetError";
    this._tag = "NoForkUrlSetError";
  }
};

// src/input/InvalidAbiError.js
var InvalidAbiError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidAbiError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidAbiErrorParameters} [args={}] - Additional parameters for the InvalidAbiError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidabierror/"
      },
      "InvalidAbiError"
    );
    this.name = "InvalidAbiError";
    this._tag = "InvalidAbiError";
  }
};

// src/input/InvalidAddToBlockchainError.js
var InvalidAddToBlockchainError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidAddToBlockchainError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidAddToBlockchainErrorParameters} [args={}] - Additional parameters for the InvalidAddToBlockchainError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidaddtoblockchainerror/"
      },
      "InvalidAddToBlockchainError"
    );
    this.name = "InvalidAddToBlockchainError";
    this._tag = "InvalidAddToBlockchainError";
  }
};

// src/input/InvalidAddToMempoolError.js
var InvalidAddToMempoolError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidAddToMempoolError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidAddToMempoolErrorParameters} [args={}] - Additional parameters for the InvalidAddToMempoolError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidaddtomempoolerror/"
      },
      "InvalidAddToMempoolError"
    );
    this.name = "InvalidAddToMempoolError";
    this._tag = "InvalidAddToMempoolError";
  }
};

// src/input/InvalidArgsError.js
var InvalidArgsError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidArgsError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidArgsErrorParameters} [args={}] - Additional parameters for the InvalidArgsError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidargserror/"
      },
      "InvalidArgsError"
    );
    this.name = "InvalidArgsError";
    this._tag = "InvalidArgsError";
  }
};

// src/input/InvalidBalanceError.js
var InvalidBalanceError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidBalanceError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidBalanceErrorParameters} [args={}] - Additional parameters for the InvalidBalanceError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidbalanceerror/"
      },
      "InvalidBalanceError"
    );
    this.name = "InvalidBalanceError";
    this._tag = "InvalidBalanceError";
  }
};

// src/input/InvalidBlobVersionedHashesError.js
var InvalidBlobVersionedHashesError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidBlobVersionedHashesError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidBlobVersionedHashesErrorParameters} [args={}] - Additional parameters for the InvalidBlobVersionedHashesError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidblobversionedhasheserror/"
      },
      "InvalidBlobVersionedHashesError"
    );
    this.name = "InvalidBlobVersionedHashesError";
    this._tag = "InvalidBlobVersionedHashesError";
  }
};

// src/input/InvalidBlockError.js
var InvalidBlockError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidBlockError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidBlockErrorParameters} [args={}] - Additional parameters for the InvalidBlockError.
   * @param {string} [tag='InvalidBlockError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidBlockError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidblockerror/"
      },
      tag
    );
  }
};

// src/input/InvalidBytecodeError.js
var InvalidBytecodeError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidBytecodeError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidBytecodeErrorParameters} [args={}] - Additional parameters for the InvalidBytecodeError.
   * @param {string} [tag='InvalidBytecodeError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidBytecodeError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidbytecodeerror/"
      },
      tag
    );
  }
};

// src/input/InvalidCallerError.js
var InvalidCallerError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidCallerError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidCallerErrorParameters} [args={}] - Additional parameters for the InvalidCallerError.
   * @param {string} [tag='InvalidCallerError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidCallerError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidcallererror/"
      },
      tag
    );
  }
};

// src/input/InvalidDataError.js
var InvalidDataError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidDataError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidDataErrorParameters} [args={}] - Additional parameters for the InvalidDataError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invaliddataerror/"
      },
      "InvalidDataError"
    );
    this.name = "InvalidDataError";
    this._tag = "InvalidDataError";
  }
};

// src/input/InvalidDeployedBytecodeError.js
var InvalidDeployedBytecodeError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidDeployedBytecodeError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidDeployedBytecodeErrorParameters} [args={}] - Additional parameters for the InvalidDeployedBytecodeError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invaliddeployedbytecodeerror/"
      },
      "InvalidDeployedBytecodeError"
    );
    this.name = "InvalidDeployedBytecodeError";
    this._tag = "InvalidDeployedBytecodeError";
  }
};

// src/input/InvalidDepthError.js
var InvalidDepthError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidDepthError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidDepthErrorParameters} [args={}] - Additional parameters for the InvalidDepthError.
   * @param {string} [tag='InvalidDepthError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidDepthError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invaliddeptherror/"
      },
      tag
    );
  }
};

// src/input/InvalidFunctionNameError.js
var InvalidFunctionNameError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidFunctionNameError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidFunctionNameErrorParameters} [args={}] - Additional parameters for the InvalidFunctionNameError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidfunctionnameerror/"
      },
      "InvalidFunctionNameError"
    );
    this.name = "InvalidFunctionNameError";
    this._tag = "InvalidFunctionNameError";
  }
};

// src/input/InvalidGasLimitError.js
var InvalidGasLimitError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidGasLimitError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidGasLimitErrorParameters} [args={}] - Additional parameters for the InvalidGasLimitError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidgaslimiterror/"
      },
      "InvalidGasLimitError"
    );
    this.name = "InvalidGasLimitError";
    this._tag = "InvalidGasLimitError";
  }
};

// src/input/InvalidGasRefundError.js
var InvalidGasRefundError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidGasRefundError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidGasRefundErrorParameters} [args={}] - Additional parameters for the InvalidGasRefundError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidgasrefunderror/"
      },
      "InvalidGasRefundError"
    );
    this.name = "InvalidGasRefundError";
    this._tag = "InvalidGasRefundError";
  }
};

// src/input/InvalidMaxFeePerGaserror.js
var InvalidMaxFeePerGasError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidMaxFeePerGasError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidMaxFeePerGasErrorParameters} [args={}] - Additional parameters for the InvalidMaxFeePerGasError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidmaxfeepergaserror/"
      },
      "InvalidMaxFeePerGasError"
    );
    this.name = "InvalidMaxFeePerGasError";
    this._tag = "InvalidMaxFeePerGasError";
  }
};

// src/input/InvalidMaxPriorityFeePerGas.js
var InvalidMaxPriorityFeePerGasError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidMaxPriorityFeePerGasError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidMaxPriorityFeePerGasErrorParameters} [args={}] - Additional parameters for the InvalidMaxPriorityFeePerGasError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidmaxpriorityfeepergaserror/"
      },
      "InvalidMaxPriorityFeePerGasError"
    );
    this.name = "InvalidMaxPriorityFeePerGasError";
    this._tag = "InvalidMaxPriorityFeePerGasError";
  }
};

// src/input/InvalidNonceError.js
var InvalidNonceError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidNonceError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidNonceErrorParameters} [args={}] - Additional parameters for the InvalidNonceError.
   * @param {string} [tag='InvalidNonceError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidNonceError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidnonceerror/"
      },
      tag
    );
  }
};

// src/input/InvalidOriginError.js
var InvalidOriginError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidOriginError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidOriginErrorParameters} [args={}] - Additional parameters for the InvalidOriginError.
   * @param {string} [tag='InvalidOriginError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidOriginError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidoriginerror/"
      },
      tag
    );
  }
};

// src/input/InvalidSaltError.js
var InvalidSaltError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidSaltError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidSaltErrorParameters} [args={}] - Additional parameters for the InvalidSaltError.
   * @param {string} [tag='InvalidSaltError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidSaltError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidsalterror/"
      },
      tag
    );
  }
};

// src/input/InvalidSelfdestructError.js
var InvalidSelfdestructError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidSelfdestructError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidSelfdestructErrorParameters} [args={}] - Additional parameters for the InvalidSelfdestructError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidselfdestructerror/"
      },
      "InvalidSelfdestructError"
    );
    this.name = "InvalidSelfdestructError";
    this._tag = "InvalidSelfdestructError";
  }
};

// src/input/InvalidSkipBalanceError.js
var InvalidSkipBalanceError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidSkipBalanceError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidSkipBalanceErrorParameters} [args={}] - Additional parameters for the InvalidSkipBalanceError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidskipbalanceerror/"
      },
      "InvalidSkipBalanceError"
    );
    this.name = "InvalidSkipBalanceError";
    this._tag = "InvalidSkipBalanceError";
  }
};

// src/input/InvalidStorageRootError.js
var InvalidStorageRootError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidStorageRootError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidStorageRootErrorParameters} [args={}] - Additional parameters for the InvalidStorageRootError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidstoragerooterror/"
      },
      "InvalidStorageRootError"
    );
    this.name = "InvalidStorageRootError";
    this._tag = "InvalidStorageRootError";
  }
};

// src/input/InvalidToError.js
var InvalidToError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidToError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidToErrorParameters} [args={}] - Additional parameters for the InvalidToError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidtoerror/"
      },
      "InvalidToError"
    );
    this.name = "InvalidToError";
    this._tag = "InvalidToError";
  }
};

// src/input/InvalidUrlError.js
var InvalidUrlError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidUrlError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidUrlErrorParameters} [args={}] - Additional parameters for the InvalidUrlError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/invalidurlerror/"
      },
      "InvalidUrlError"
    );
    this.name = "InvalidUrlError";
    this._tag = "InvalidUrlError";
  }
};

// src/input/InvalidValueError.js
var InvalidValueError = class extends InvalidParamsError {
  /**
   * Constructs an InvalidValueError.
   *
   * @param {string} message - Human-readable error message.
   * @param {InvalidValueErrorParameters} [args={}] - Additional parameters for the InvalidValueError.
   * @param {string} [tag='InvalidValueError'] - The tag for the error.
   */
  constructor(message, args = {}, tag = "InvalidValueError") {
    super(
      message,
      {
        ...args,
        docsBaseUrl: "https://tevm.sh",
        docsPath: "/reference/tevm/errors/classes/invalidvalueerror/"
      },
      tag
    );
  }
};

// src/rpcErrorToMessage.js
var rpcErrorCodeToMessage = {
  [-1]: "Unknown error",
  // Standard JSON-RPC 2.0 errors
  [-32700]: "Parse error",
  [-32600]: "Invalid request",
  [-32601]: "Method not found",
  [-32602]: "Invalid params",
  [-32603]: "Internal error",
  // EIP-1474 and common convention errors
  [-32e3]: "Invalid input",
  [-32001]: "Resource not found",
  [-32002]: "Resource unavailable",
  [-32003]: "Transaction rejected",
  [-32004]: "Method not supported",
  [-32005]: "Limit exceeded",
  [-32015]: "VM execution error",
  [-32020]: "Account locked",
  [3]: "Execution reverted",
  // Tevm-specific implementation errors
  [-32006]: "Block gas limit exceeded",
  [-32007]: "Unsupported chain",
  [-32008]: "Nonce already used",
  [-32009]: "Insufficient permissions",
  [-32011]: "Transaction too large",
  [-32012]: "Invalid gas price",
  [-32013]: "Invalid address",
  [-32014]: "Transaction underpriced",
  // Other provider-specific errors (for compatibility)
  [4001]: "User rejected request",
  [4100]: "Unauthorized",
  [4200]: "Method not supported",
  [4900]: "Provider disconnected",
  [4901]: "Provider disconnected from chain",
  [-32010]: "Transaction cost exceeds current gas limit",
  [-32042]: "Method not found",
  [-32500]: "Invalid user operation",
  [-32501]: "User operation validation error",
  [-32521]: "Invalid signature",
  [-32604]: "Record not found",
  [-32605]: "Rate Limit",
  [-32606]: "Illegal Ip",
  [-32607]: "Call Limit",
  [-32612]: "Tracing/debugging blocked",
  [-32613]: "Tracing/debugging blocked",
  [-32616]: "WebSocket response size exceeded",
  [-33e3]: "Payment required",
  [-33100]: "Forbidden",
  [-33200]: "Too Many Requests",
  [-33300]: "Too Many Requests",
  [-33400]: "Too Many Requests"
};

// src/utils/DecodeFunctionDataError.js
var DecodeFunctionDataError = class extends InvalidParamsError {
  /**
   * Constructs a DecodeFunctionDataError.
   *
   * @param {string} message - Human-readable error message.
   * @param {DecodeFunctionDataErrorParameters} [args={}] - Additional parameters for the DecodeFunctionDataError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/decodefunctiondataerror/"
      },
      "DecodeFunctionDataError"
    );
    this.name = "DecodeFunctionDataError";
    this._tag = "DecodeFunctionDataError";
  }
};

// src/utils/EncodeFunctionReturnDataError.js
var EncodeFunctionReturnDataError = class extends InvalidParamsError {
  /**
   * Constructs an EncodeFunctionReturnDataError.
   *
   * @param {string} message - Human-readable error message.
   * @param {EncodeFunctionReturnDataErrorParameters} [args={}] - Additional parameters for the EncodeFunctionReturnDataError.
   */
  constructor(message, args = {}) {
    super(
      message,
      {
        ...args,
        docsBaseUrl: args.docsBaseUrl ?? "https://tevm.sh",
        docsPath: args.docsPath ?? "/reference/tevm/errors/classes/encodefunctionreturndataerror/"
      },
      "EncodeFunctionReturnDataError"
    );
    this.name = "EncodeFunctionReturnDataError";
    this._tag = "EncodeFunctionReturnDataError";
  }
};

Object.defineProperty(exports, "InternalRpcError", {
  enumerable: true,
  get: function () { return viem.InternalRpcError; }
});
Object.defineProperty(exports, "InvalidInputRpcError", {
  enumerable: true,
  get: function () { return viem.InvalidInputRpcError; }
});
Object.defineProperty(exports, "InvalidParamsRpcError", {
  enumerable: true,
  get: function () { return viem.InvalidParamsRpcError; }
});
Object.defineProperty(exports, "InvalidRequestRpcError", {
  enumerable: true,
  get: function () { return viem.InvalidRequestRpcError; }
});
Object.defineProperty(exports, "JsonRpcVersionUnsupportedError", {
  enumerable: true,
  get: function () { return viem.JsonRpcVersionUnsupportedError; }
});
Object.defineProperty(exports, "LimitExceededRpcError", {
  enumerable: true,
  get: function () { return viem.LimitExceededRpcError; }
});
Object.defineProperty(exports, "MethodNotFoundRpcError", {
  enumerable: true,
  get: function () { return viem.MethodNotFoundRpcError; }
});
Object.defineProperty(exports, "MethodNotSupportedRpcError", {
  enumerable: true,
  get: function () { return viem.MethodNotSupportedRpcError; }
});
Object.defineProperty(exports, "ParseRpcError", {
  enumerable: true,
  get: function () { return viem.ParseRpcError; }
});
Object.defineProperty(exports, "ProviderDisconnectedError", {
  enumerable: true,
  get: function () { return viem.ProviderDisconnectedError; }
});
Object.defineProperty(exports, "ResourceNotFoundRpcError", {
  enumerable: true,
  get: function () { return viem.ResourceNotFoundRpcError; }
});
Object.defineProperty(exports, "ResourceUnavailableRpcError", {
  enumerable: true,
  get: function () { return viem.ResourceUnavailableRpcError; }
});
Object.defineProperty(exports, "TimeoutError", {
  enumerable: true,
  get: function () { return viem.TimeoutError; }
});
Object.defineProperty(exports, "UnknownRpcError", {
  enumerable: true,
  get: function () { return viem.UnknownRpcError; }
});
Object.defineProperty(exports, "UnsupportedProviderMethodError", {
  enumerable: true,
  get: function () { return viem.UnsupportedProviderMethodError; }
});
Object.defineProperty(exports, "UserRejectedRequestError", {
  enumerable: true,
  get: function () { return viem.UserRejectedRequestError; }
});
exports.AccountLockedError = AccountLockedError;
exports.AccountNotFoundError = AccountNotFoundError;
exports.AuthCallUnsetError = AuthCallUnsetError;
exports.BLS12381FpNotInFieldError = BLS12381FpNotInFieldError;
exports.BLS12381InputEmptyError = BLS12381InputEmptyError;
exports.BLS12381InvalidInputLengthError = BLS12381InvalidInputLengthError;
exports.BLS12381PointNotOnCurveError = BLS12381PointNotOnCurveError;
exports.BaseError = BaseError;
exports.BlobGasLimitExceededError = BlobGasLimitExceededError;
exports.BlockGasLimitExceededError = BlockGasLimitExceededError;
exports.ChainIdMismatchError = ChainIdMismatchError;
exports.CodeSizeExceedsMaximumError = CodeSizeExceedsMaximumError;
exports.CodeStoreOutOfGasError = CodeStoreOutOfGasError;
exports.CommonMismatchError = CommonMismatchError;
exports.ContractExecutionFailedError = ContractExecutionFailedError;
exports.CreateCollisionError = CreateCollisionError;
exports.DecodeFunctionDataError = DecodeFunctionDataError;
exports.DefensiveNullCheckError = DefensiveNullCheckError;
exports.EipNotEnabledError = EipNotEnabledError;
exports.EncodeFunctionReturnDataError = EncodeFunctionReturnDataError;
exports.EvmRevertError = EvmRevertError;
exports.ExecutionError = ExecutionError;
exports.ForkError = ForkError;
exports.GasLimitExceededError = GasLimitExceededError;
exports.InitcodeSizeViolationError = InitcodeSizeViolationError;
exports.InsufficientBalanceError = InsufficientBalanceError;
exports.InsufficientFundsError = InsufficientFundsError;
exports.InsufficientPermissionsError = InsufficientPermissionsError;
exports.InternalError = InternalError;
exports.InternalEvmError = InternalEvmError;
exports.InvalidAbiError = InvalidAbiError;
exports.InvalidAddToBlockchainError = InvalidAddToBlockchainError;
exports.InvalidAddToMempoolError = InvalidAddToMempoolError;
exports.InvalidAddressError = InvalidAddressError;
exports.InvalidArgsError = InvalidArgsError;
exports.InvalidBalanceError = InvalidBalanceError;
exports.InvalidBeginSubError = InvalidBeginSubError;
exports.InvalidBlobVersionedHashesError = InvalidBlobVersionedHashesError;
exports.InvalidBlockError = InvalidBlockError;
exports.InvalidBytecodeError = InvalidBytecodeError;
exports.InvalidBytecodeResultError = InvalidBytecodeResultError;
exports.InvalidBytesSizeError = InvalidBytesSizeError;
exports.InvalidCallerError = InvalidCallerError;
exports.InvalidCommitmentError = InvalidCommitmentError;
exports.InvalidDataError = InvalidDataError;
exports.InvalidDeployedBytecodeError = InvalidDeployedBytecodeError;
exports.InvalidDepthError = InvalidDepthError;
exports.InvalidEofFormatError = InvalidEofFormatError;
exports.InvalidFunctionNameError = InvalidFunctionNameError;
exports.InvalidGasLimitError = InvalidGasLimitError;
exports.InvalidGasPriceError = InvalidGasPriceError;
exports.InvalidGasRefundError = InvalidGasRefundError;
exports.InvalidInputLengthError = InvalidInputLengthError;
exports.InvalidJumpError = InvalidJumpError;
exports.InvalidJumpSubError = InvalidJumpSubError;
exports.InvalidKzgInputsError = InvalidKzgInputsError;
exports.InvalidMaxFeePerGasError = InvalidMaxFeePerGasError;
exports.InvalidMaxPriorityFeePerGasError = InvalidMaxPriorityFeePerGasError;
exports.InvalidNonceError = InvalidNonceError;
exports.InvalidOpcodeError = InvalidOpcodeError;
exports.InvalidOriginError = InvalidOriginError;
exports.InvalidParamsError = InvalidParamsError;
exports.InvalidProofError = InvalidProofError;
exports.InvalidRequestError = InvalidRequestError;
exports.InvalidReturnSubError = InvalidReturnSubError;
exports.InvalidSaltError = InvalidSaltError;
exports.InvalidSelfdestructError = InvalidSelfdestructError;
exports.InvalidSignatureError = InvalidSignatureError;
exports.InvalidSkipBalanceError = InvalidSkipBalanceError;
exports.InvalidStorageRootError = InvalidStorageRootError;
exports.InvalidToError = InvalidToError;
exports.InvalidTransactionError = InvalidTransactionError;
exports.InvalidUrlError = InvalidUrlError;
exports.InvalidValueError = InvalidValueError;
exports.LimitExceededError = LimitExceededError;
exports.MethodNotFoundError = MethodNotFoundError;
exports.MethodNotSupportedError = MethodNotSupportedError;
exports.MisconfiguredClientError = MisconfiguredClientError;
exports.NoForkTransportSetError = NoForkTransportSetError;
exports.NoForkUrlSetError = NoForkUrlSetError;
exports.NonceAlreadyUsedError = NonceAlreadyUsedError;
exports.NonceTooHighError = NonceTooHighError;
exports.NonceTooLowError = NonceTooLowError;
exports.OutOfGasError = OutOfGasError;
exports.OutOfRangeError = OutOfRangeError;
exports.ParseError = ParseError;
exports.PendingTransactionTimeoutError = PendingTransactionTimeoutError;
exports.RateLimitExceededError = RateLimitExceededError;
exports.RefundExhaustedError = RefundExhaustedError;
exports.ResourceNotFoundError = ResourceNotFoundError;
exports.ResourceUnavailableError = ResourceUnavailableError;
exports.RevertError = RevertError;
exports.StackOverflowError = StackOverflowError;
exports.StackUnderflowError = StackUnderflowError;
exports.StaticStateChangeError = StaticStateChangeError;
exports.StopError = StopError;
exports.TransactionRejectedError = TransactionRejectedError;
exports.TransactionTooLargeError = TransactionTooLargeError;
exports.TransactionUnderpricedError = TransactionUnderpricedError;
exports.UnknownBlockError = UnknownBlockError;
exports.UnreachableCodeError = UnreachableCodeError;
exports.UnsupportedChainError = UnsupportedChainError;
exports.ValueOverflowError = ValueOverflowError;
exports.rpcErrorCodeToMessage = rpcErrorCodeToMessage;
