// @tevm/jsonrpc/dist/index.js
export { loadBalance, rateLimit } from '@ponder/utils';
export { http, webSocket } from 'viem';

// src/index.ts

// src/createJsonRpcFetcher.js
var createJsonRpcFetcher = (client) => {
  return {
    request: async (request) => {
      try {
        const result = await client.request(request);
        return {
          jsonrpc: "2.0",
          result,
          method: request.method,
          ...request.params !== void 0 ? { params: request.params } : {},
          ...request.id !== void 0 ? { id: request.id } : {}
        };
      } catch (e) {
        if (typeof e === "object" && e !== null && "code" in e) {
          return {
            jsonrpc: "2.0",
            error: {
              code: (
                /** @type {string}*/
                e.code
              ),
              message: "message" in e ? (
                /** @type {string}*/
                e.message
              ) : "Unknown error in jsonrpc request"
            },
            method: request.method,
            ...request.params !== void 0 ? { params: request.params } : {},
            ...request.id !== void 0 ? { id: request.id } : {}
          };
        }
        return {
          jsonrpc: "2.0",
          error: {
            code: -32e3,
            message: "Unknown error in jsonrpc request"
          },
          method: request.method,
          ...request.params !== void 0 ? { params: request.params } : {},
          ...request.id !== void 0 ? { id: request.id } : {}
        };
      }
    }
  };
};

export { createJsonRpcFetcher };
