// @tevm/jsonrpc/dist/index.cjs
'use strict';

var utils = require('@ponder/utils');
var viem = require('viem');

// src/index.ts

// src/createJsonRpcFetcher.js
var createJsonRpcFetcher = (client) => {
  return {
    request: async (request) => {
      try {
        const result = await client.request(request);
        return {
          jsonrpc: "2.0",
          result,
          method: request.method,
          ...request.params !== void 0 ? { params: request.params } : {},
          ...request.id !== void 0 ? { id: request.id } : {}
        };
      } catch (e) {
        if (typeof e === "object" && e !== null && "code" in e) {
          return {
            jsonrpc: "2.0",
            error: {
              code: (
                /** @type {string}*/
                e.code
              ),
              message: "message" in e ? (
                /** @type {string}*/
                e.message
              ) : "Unknown error in jsonrpc request"
            },
            method: request.method,
            ...request.params !== void 0 ? { params: request.params } : {},
            ...request.id !== void 0 ? { id: request.id } : {}
          };
        }
        return {
          jsonrpc: "2.0",
          error: {
            code: -32e3,
            message: "Unknown error in jsonrpc request"
          },
          method: request.method,
          ...request.params !== void 0 ? { params: request.params } : {},
          ...request.id !== void 0 ? { id: request.id } : {}
        };
      }
    }
  };
};

Object.defineProperty(exports, "loadBalance", {
  enumerable: true,
  get: function () { return utils.loadBalance; }
});
Object.defineProperty(exports, "rateLimit", {
  enumerable: true,
  get: function () { return utils.rateLimit; }
});
Object.defineProperty(exports, "http", {
  enumerable: true,
  get: function () { return viem.http; }
});
Object.defineProperty(exports, "webSocket", {
  enumerable: true,
  get: function () { return viem.webSocket; }
});
exports.createJsonRpcFetcher = createJsonRpcFetcher;
